---
title: "taxa_DESeq2_phyloseq"
author: "Zeya Zhengyao Xue"
date: "October 8, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

I used old .Rmd and html with rarefied feature table to generate alpha and beta 
diversity figures. This script is focused on differential abudance testing on 
taxonomy profile.

## Setting up
```{r}
library(phyloseq);packageVersion("phyloseq")
library(DESeq2);packageVersion("DESeq2")
#library(ggplot2);packageVersion("ggplot2")
#library(reshape2)
#library(cowplot)
library(plyr)
library(dplyr)
#library(RColorBrewer)
#library(microbiomeSeq)

# Define project path
path <- "G:/My Drive/UC_Davis/Marco_lab/milk_microbiota/Hilmar_weekly_samples/QIIME2/"
# copy the .Rmd file to the path directory to generate html report

set.seed(123) #  for reproducibility

# remake a unrarefied ps object with low coverage samples filtered
## The starting ps is removed of ASV with undetermined phylum and ASV < 0.00005
ps05.Phy05 <- readRDS(file.path(path, "decontam/phyloseq/ps05.Phy05.rds"))
ps05.Phy05
# add tree
tree <- read_tree(file.path(path, "decontam/phyloseq/tree.nwk"))
ps05.Phy05.tre <- merge_phyloseq(ps05.Phy05, phy_tree(tree))
saveRDS(ps05.Phy05.tre, file.path(path, "decontam/phyloseq/ps05.Phy05.tre.rds"))
ps05.Phy05.tre

# remove sample < 100 reads
ps100 <- prune_samples(sample_sums(ps05.Phy05.tre) >= 100, ps05.Phy05.tre)
saveRDS(ps100, file.path(path, "decontam/phyloseq/ps100.rds"))
ps100

# remove all NC and seq control samples 
ps100.sam <- subset_samples(ps100, SampleOrCtrl == "Sample" )
#ps100.sam <- ps100.sam %>% tax_glom(taxrank = "Genus")
saveRDS(ps100.sam, file.path(path, "decontam/phyloseq/ps100.sam.rds"))
ps100.sam
```


## Define function to use DESeq2 for differential analysis 
```{r}
# calculate geometric means prior to estimate size factors
gmMean = function(x, na.rm=TRUE){
  exp(sum(log(x[x > 0]), na.rm=na.rm) / length(x))
}

# Differential abundance and generate diagnostic plots and write sig tables 
## fold change is slits vs no slits
## Uses default Benjamini-Hochberg pvalue adjust
DA <- function(ps, path.out, variable = c("CheeseOutcome","TimeSinceCleaning")) {
  psdds <- phyloseq_to_deseq2(ps, ~ variable)
  geoMeans <- apply(counts(psdds), 1, gmMean)
  psdds <- estimateSizeFactors(psdds, geoMeans = geoMeans)
  dds <- DESeq(psdds, test = "Wald", fitType = "local") 
  plotDispEsts(dds, ylim = c(1e-6, 1e2)) 
  alpha <- 0.1 #  padj for indepenent filtering and expect FDR < alpha
  res <- results(dds, alpha = alpha, 
                 lfcThreshold = log2(1.5), 
                 altHypothesis = "greaterAbs") #  turn on outlier detection
  mcols(res, use.names=TRUE)  
  plotMA(res) # diagnostic description
  hist(res$pvalue, breaks=20, col="grey" ) # diagnostic plot
  sigtab = res[which(res$padj < alpha), ] # diagnostic plot
  if (nrow(sigtab) == 0) {
    print("There are no significant features whose padj < 0.1 and |lfc| > log2(1.5)")
  } else {
    sigtab <- cbind(as(sigtab, "data.frame"), as(tax_table(ps)[rownames(sigtab), ], "matrix"))
    write.csv(sigtab, file.path(path.out))
  }
}
```


## Taxonomic DA of Cheddar 
This is PMA treated samples
```{r}
# 1 # All Cheddar samples 
ps.ched <- subset_samples(ps100.sam, CheeseType %in% "Cheddar")
ps.ched <- subset_samples(ps.ched, 
                          SampleType %in% c("Cheese_0D","Cheese_30D","Cheese_90D","Cheese_120D"))
saveRDS(ps.ched, file.path(path, "cheddar/ps.ched.rds"))
ps.ched
# PMA 
ps.chedY <- subset_samples(ps.ched, PMA %in% "Y")
saveRDS(ps.chedY, file.path(path, "cheddar/ps.chedY.rds"))
ps.chedY
CheeseOutcomeDA(ps.chedY, path.out = file.path(path, "cheddar/deseq_table/sigtab.chedY.csv"))

# 2 # All Cheddar samples without Lactococcus
ps.chedY.noL <- subset_taxa(ps.chedY, !Genus %in% "Lactococcus")
saveRDS(ps.chedY.noL, file.path(path, "cheddar/ps.chedY.noL.rds"))
ps.chedY.noL
CheeseOutcomeDA(ps.chedY.noL, 
                path.out = file.path(path, "cheddar/deseq_table/sigtab.chedY.noL.csv"))

# 3 # 0-day Cheddar samples without Lactococcus
ps.chedY.noL0 <- subset_samples(ps.chedY.noL, SampleType %in% "Cheese_0D")
saveRDS(ps.chedY.noL0, file.path(path, "cheddar/ps.chedY.noL0.rds"))
ps.chedY.noL0
CheeseOutcomeDA(ps.chedY.noL0, 
                path.out = file.path(path, "cheddar/deseq_table/sigtab.chedY.noL0.csv"))

# 4 # 30-day Cheddar samples without Lactococcus
ps.chedY.noL30 <- subset_samples(ps.chedY.noL, SampleType %in% "Cheese_30D")
saveRDS(ps.chedY.noL30, file.path(path, "cheddar/ps.chedY.noL30.rds"))
ps.chedY.noL30
CheeseOutcomeDA(ps.chedY.noL30, 
                path.out = file.path(path, "cheddar/deseq_table/sigtab.chedY.noL30.csv"))

# 5 # 90-day Cheddar samples without Lactococcus
ps.chedY.noL90 <- subset_samples(ps.chedY.noL, SampleType %in% "Cheese_90D")
saveRDS(ps.chedY.noL90, file.path(path, "cheddar/ps.chedY.noL90.rds"))
ps.chedY.noL90
CheeseOutcomeDA(ps.chedY.noL90, 
                path.out = file.path(path, "cheddar/deseq_table/sigtab.chedY.noL90.csv"))

# 6 # 120-day Cheddar samples without Lactococcus
ps.chedY.noL120 <- subset_samples(ps.chedY.noL, SampleType %in% "Cheese_120D")
saveRDS(ps.chedY.noL120, file.path(path, "cheddar/ps.chedY.noL120.rds"))
ps.chedY.noL120
CheeseOutcomeDA(ps.chedY.noL120, 
                path.out = file.path(path, "cheddar/deseq_table/sigtab.chedY.noL120.csv"))
```

## Taxonomic DA of HTST milk that has been adjusted for cell count using qPCR results
This is PMA treated samples
```{r}
# 7 # HTST milk 
ps.htst <- subset_samples(ps100.sam, SampleType %in% "HTST_milk")
saveRDS(ps.htst, file.path(path, "cheddar/ps.htst.rds"))
ps.htst
ps.htstY <- subset_samples(ps.htst, PMA %in% "Y")
saveRDS(ps.htstY, file.path(path, "cheddar/ps.htstY.rds"))
ps.htstY
CheeseOutcomeDA(ps.htstY, 
                path.out = file.path(path, "milk/deseq_table/sigtab.htstY.csv"))

# 8 # HTST milk that are 
```


## Taxonomic analysis of each sample type 
This is PMA untreated samples
```{r}

```

