---
title: "taxa_DESeq2_phyloseq"
author: "Zeya Zhengyao Xue"
date: "October 8, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


I used old .Rmd and html with rarefied feature table to generate alpha and beta 
diversity figures. This script is focused on differential abudance testing on 
taxonomy profile.

## Setting up
```{r}
library(phyloseq);packageVersion("phyloseq")
library(DESeq2);packageVersion("DESeq2")
library(ggplot2)
library(reshape2)
library(superheat)
library(plyr)
library(dplyr)


# Define project path
path <- "G:/My Drive/UC_Davis/Marco_lab/milk_microbiota/Hilmar_weekly_samples/QIIME2/"
# copy the .Rmd file to the path directory to generate html report

set.seed(123) #  for reproducibility

# character vector for the comm 14 most abundant taxa
comm14 <- c("Acinetobacter", "Bacillus", "Corynebacterium", "Clostridium",
            "Enterococcus", "Lactobacillus", "Lactococcus", "Macrococcus",
            "Mycoplasma", "Pseudomonas", "Staphylococcus", "Streptococcus",
            "Thermus", "Turicibacter", "Other")
ched14 <- c("Acinetobacter", "Bacillus", "Corynebacterium", "Clostridium",
            "Enterococcus", "Leuconostoc", "Lactobacillus", "Macrococcus",
            "Mycoplasma", "Pseudomonas", "Staphylococcus", "Streptococcus",
            "Thermus", "Turicibacter", "Other")
comm18 <- c("Acinetobacter", "Bacillus", "Brevibacillus", "Corynebacterium", 
            "Clostridium", "Enterococcus", "Escherichia", "Geobacillus",
            "Lactobacillus", "Lactococcus", "Leuconostoc", "Macrococcus", 
            "Mycoplasma", "Pseudomonas", "Staphylococcus", "Streptococcus",
            "Thermus", "Turicibacter", "Other")
```


## Define function to use DESeq2 for differential analysis 
```{r}
# calculate geometric means prior to estimate size factors
gmMean = function(x, na.rm=TRUE){
  exp(sum(log(x[x > 0]), na.rm=na.rm) / length(x))
}

# Cheese Outcome Differential abundance and generate diagnostic plots and write sig tables 
## fold change is slits vs no slits
## Uses default Benjamini-Hochberg pvalue adjust
CheeseOutcomeDA <- function(ps, path.out) {
  ps <- ps %>% tax_glom(taxrank = "Genus")
  psdds <- phyloseq_to_deseq2(ps, ~ CheeseOutcome)
  
  geoMeans <- apply(counts(psdds), 1, gmMean)
  psdds <- estimateSizeFactors(psdds, geoMeans = geoMeans)
  
  dds <- DESeq(psdds, test = "Wald", fitType = "local") 
  plotDispEsts(dds, ylim = c(1e-6, 1e2)) 
  alpha <- 0.1 #  padj for indepenent filtering and expect FDR < alpha
  res <- results(dds, alpha = alpha, 
                 lfcThreshold = 0, 
                 altHypothesis = "greaterAbs")
  mcols(res, use.names=TRUE)  
  plotMA(res) # diagnostic description
  hist(res$pvalue, breaks=20, col="grey" ) # diagnostic plot
  sigtab = res[which(res$padj < alpha), ] 
  if (nrow(sigtab) == 0) {
    print("There are no significant features whose padj < 0.1 and |lfc| > 0")
  } else {
    sigtab <- cbind(as(sigtab, "data.frame"), as(tax_table(ps)[rownames(sigtab), ], "matrix"))
    write.csv(sigtab, file.path(path.out))
  }
}

CheeseOutcomeDAnoMean <- function(ps, path.out) {
  ps <- ps %>% tax_glom(taxrank = "Genus")
  psdds <- phyloseq_to_deseq2(ps, ~ CheeseOutcome)
  
  dds <- DESeq(psdds, test = "Wald", fitType = "local") 
  plotDispEsts(dds, ylim = c(1e-6, 1e2)) 
  alpha <- 0.5 #  padj for indepenent filtering and expect FDR < alpha
  res <- results(dds, alpha = alpha, 
                 lfcThreshold = 0, 
                 altHypothesis = "greaterAbs")
  mcols(res, use.names=TRUE)  
  plotMA(res) # diagnostic description
  hist(res$pvalue, breaks=20, col="grey" ) # diagnostic plot
  sigtab = res[which(res$padj < alpha), ] 
  if (nrow(sigtab) == 0) {
    print("There are no significant features whose padj < 0.5 and |lfc| > 0")
  } else {
    sigtab <- cbind(as(sigtab, "data.frame"), as(tax_table(ps)[rownames(sigtab), ], "matrix"))
    write.csv(sigtab, file.path(path.out))
  }
}

# TimeSinceCleaning Differential abundance and generate diagnostic plots and write sig tables 
TiSinCleaningDA <- function(ps, path.out) {
  ps <- ps %>% tax_glom(taxrank = "Genus")
  sample_data(ps)$TimeSinceCleaning <- factor(sample_data(ps)$TimeSinceCleaning)
  psdds <- phyloseq_to_deseq2(ps, design = ~ TimeSinceCleaning)
  geoMeans <- apply(counts(psdds), 1, gmMean)
  psdds <- estimateSizeFactors(psdds, geoMeans = geoMeans)
  
  dds <- DESeq(psdds, test = "Wald", fitType = "local") 
  plotDispEsts(dds, ylim = c(1e-6, 1e2)) 
  alpha <- 0.5 #  padj for indepenent filtering and expect FDR < alpha

  res03 <- results(dds, contrast = c("TimeSinceCleaning", "0", "3"),
                   alpha = alpha, lfcThreshold = 0, altHypothesis = "greaterAbs")
  res06 <- results(dds, contrast = c("TimeSinceCleaning", "0", "6"),
                   alpha = alpha, lfcThreshold = 0, altHypothesis = "greaterAbs")
  res09 <- results(dds, contrast = c("TimeSinceCleaning", "0", "9"),
                   alpha = alpha, lfcThreshold = 0, altHypothesis = "greaterAbs")
  res36 <- results(dds, contrast = c("TimeSinceCleaning", "3", "6"),
                   alpha = alpha, lfcThreshold = 0, altHypothesis = "greaterAbs")
  res39 <- results(dds, contrast = c("TimeSinceCleaning", "3", "9"),
                   alpha = alpha, lfcThreshold = 0, altHypothesis = "greaterAbs")
  res69 <- results(dds, contrast = c("TimeSinceCleaning", "6", "9"),
                   alpha = alpha, lfcThreshold = 0, altHypothesis = "greaterAbs")
  
  f <- function(res){
    mcols(res, use.names = TRUE)
    sigtab <- res[which(res$padj < alpha), ]
  }
  
  f2 <- function(sigtab){
    cbind(as(sigtab, "data.frame"), as(tax_table(ps)[rownames(sigtab), ], "matrix"))
  }
  
  sigtab03 <- f(res03)
  sigtab06 <- f(res06)
  sigtab09 <- f(res09)
  sigtab36 <- f(res36)
  sigtab39 <- f(res39)
  sigtab69 <- f(res69)
  
  if (nrow(sigtab03) > 0){
    sigtab03 <- f2(sigtab03)
    write.csv(sigtab03, file.path(path.out, "03.csv"))
  }
  if (nrow(sigtab06) > 0){
    sigtab06 <- f2(sigtab06) 
    write.csv(sigtab06, file.path(path.out, "06.csv"))
  }
  if (nrow(sigtab09) > 0){
    sigtab09 <- f2(sigtab09) 
    write.csv(sigtab09, file.path(path.out, "09.csv"))
  }
  if (nrow(sigtab36) > 0){
   sigtab36 <- f2(sigtab36) 
   write.csv(sigtab36, file.path(path.out, "36.csv"))
  }
  if (nrow(sigtab39) > 0){
    sigtab39 <- f2(sigtab39) 
    write.csv(sigtab39, file.path(path.out, "39.csv"))
  }
  if (nrow(sigtab69) > 0){
    sigtab69 <- f2(sigtab69) 
    write.csv(sigtab69, file.path(path.out, "69.csv"))
  }
  
}

TiSinCleaningDAnoMean <- function(ps, path.out) {
  ps <- ps %>% tax_glom(taxrank = "Genus")
  sample_data(ps)$TimeSinceCleaning <- factor(sample_data(ps)$TimeSinceCleaning)
  psdds <- phyloseq_to_deseq2(ps, design = ~ TimeSinceCleaning)
  
  dds <- DESeq(psdds, test = "Wald", fitType = "local") 
  plotDispEsts(dds, ylim = c(1e-6, 1e2)) 
  alpha <- 0.5 #  padj for indepenent filtering and expect FDR < alpha

  res03 <- results(dds, contrast = c("TimeSinceCleaning", "0", "3"),
                   alpha = alpha, lfcThreshold = 0, altHypothesis = "greaterAbs")
  res06 <- results(dds, contrast = c("TimeSinceCleaning", "0", "6"),
                   alpha = alpha, lfcThreshold = 0, altHypothesis = "greaterAbs")
  res09 <- results(dds, contrast = c("TimeSinceCleaning", "0", "9"),
                   alpha = alpha, lfcThreshold = 0, altHypothesis = "greaterAbs")
  res36 <- results(dds, contrast = c("TimeSinceCleaning", "3", "6"),
                   alpha = alpha, lfcThreshold = 0, altHypothesis = "greaterAbs")
  res39 <- results(dds, contrast = c("TimeSinceCleaning", "3", "9"),
                   alpha = alpha, lfcThreshold = 0, altHypothesis = "greaterAbs")
  res69 <- results(dds, contrast = c("TimeSinceCleaning", "6", "9"),
                   alpha = alpha, lfcThreshold = 0, altHypothesis = "greaterAbs")
  
  f <- function(res){
    mcols(res, use.names = TRUE)
    sigtab <- res[which(res$padj < alpha), ]
  }
  
  f2 <- function(sigtab){
    cbind(as(sigtab, "data.frame"), as(tax_table(ps)[rownames(sigtab), ], "matrix"))
  }
  
  sigtab03 <- f(res03)
  sigtab06 <- f(res06)
  sigtab09 <- f(res09)
  sigtab36 <- f(res36)
  sigtab39 <- f(res39)
  sigtab69 <- f(res69)
  
  if (nrow(sigtab03) > 0){
    sigtab03 <- f2(sigtab03)
    write.csv(sigtab03, file.path(path.out, "03.csv"))
  }
  if (nrow(sigtab06) > 0){
    sigtab06 <- f2(sigtab06) 
    write.csv(sigtab06, file.path(path.out, "06.csv"))
  }
  if (nrow(sigtab09) > 0){
    sigtab09 <- f2(sigtab09) 
    write.csv(sigtab09, file.path(path.out, "09.csv"))
  }
  if (nrow(sigtab36) > 0){
   sigtab36 <- f2(sigtab36) 
   write.csv(sigtab36, file.path(path.out, "36.csv"))
  }
  if (nrow(sigtab39) > 0){
    sigtab39 <- f2(sigtab39) 
    write.csv(sigtab39, file.path(path.out, "39.csv"))
  }
  if (nrow(sigtab69) > 0){
    sigtab69 <- f2(sigtab69) 
    write.csv(sigtab69, file.path(path.out, "69.csv"))
  }
  
}


# DA function for milk type
MilkDA <- function(ps, path.out) {
  ps <- ps %>% tax_glom(taxrank = "Genus")
  sample_data(ps)$SampleType <- factor(sample_data(ps)$SampleType)
  psdds <- phyloseq_to_deseq2(ps, design = ~ SampleType)
  geoMeans <- apply(counts(psdds), 1, gmMean)
  psdds <- estimateSizeFactors(psdds, geoMeans = geoMeans)
  
  dds <- DESeq(psdds, test = "Wald", fitType = "local") 
  plotDispEsts(dds, ylim = c(1e-6, 1e2)) 
  alpha <- 0.5 #  padj for indepenent filtering and expect FDR < alpha

  res01 <- results(dds, contrast = c("SampleType", "Raw_milk", "HTST_feed"),
                   alpha = alpha, lfcThreshold = 0, altHypothesis = "greaterAbs")
  res02 <- results(dds, contrast = c("SampleType", "Raw_milk", "HTST_milk"),
                   alpha = alpha, lfcThreshold = 0, altHypothesis = "greaterAbs")
  res12 <- results(dds, contrast = c("SampleType", "HTST_feed", "HTST_milk"),
                   alpha = alpha, lfcThreshold = 0, altHypothesis = "greaterAbs")
  
  f <- function(res){
    mcols(res, use.names = TRUE)
    sigtab <- res[which(res$padj < alpha), ]
  }
  
  f2 <- function(sigtab){
    cbind(as(sigtab, "data.frame"), as(tax_table(ps)[rownames(sigtab), ], "matrix"))
  }
  
  sigtab01 <- f(res01)
  sigtab02 <- f(res02)
  sigtab12 <- f(res12)
  
  if (nrow(sigtab01) > 0){
    sigtab01 <- f2(sigtab01)
    write.csv(sigtab01, file.path(path.out, "01.csv"))
  }
  if (nrow(sigtab02) > 0){
    sigtab02 <- f2(sigtab02) 
    write.csv(sigtab02, file.path(path.out, "02.csv"))
  }
  if (nrow(sigtab12) > 0){
    sigtab12 <- f2(sigtab12) 
    write.csv(sigtab12, file.path(path.out, "12.csv"))
  }
  
}


# DA function for Cheese type
ChedDA <- function(ps, path.out) {
  ps <- ps %>% tax_glom(taxrank = "Genus")
  sample_data(ps)$SampleType <- factor(sample_data(ps)$SampleType)
  psdds <- phyloseq_to_deseq2(ps, design = ~ SampleType)
  geoMeans <- apply(counts(psdds), 1, gmMean)
  psdds <- estimateSizeFactors(psdds, geoMeans = geoMeans)
  
  dds <- DESeq(psdds, test = "Wald", fitType = "local") 
  plotDispEsts(dds, ylim = c(1e-6, 1e2)) 
  alpha <- 0.5 #  padj for indepenent filtering and expect FDR < alpha

  res03 <- results(dds, contrast = c("SampleType", "Cheese_0D", "Cheese_30D"),
                   alpha = alpha, lfcThreshold = 0, altHypothesis = "greaterAbs")
  res09 <- results(dds, contrast = c("SampleType", "Cheese_0D", "Cheese_90D"),
                   alpha = alpha, lfcThreshold = 0, altHypothesis = "greaterAbs")
  res012 <- results(dds, contrast = c("SampleType", "Cheese_0D", "Cheese_120D"),
                   alpha = alpha, lfcThreshold = 0, altHypothesis = "greaterAbs")
  res39 <- results(dds, contrast = c("SampleType", "Cheese_30D", "Cheese_90D"),
                   alpha = alpha, lfcThreshold = 0, altHypothesis = "greaterAbs")
  res312 <- results(dds, contrast = c("SampleType", "Cheese_30D", "Cheese_120D"),
                   alpha = alpha, lfcThreshold = 0, altHypothesis = "greaterAbs")
  res912 <- results(dds, contrast = c("SampleType", "Cheese_90D", "Cheese_120D"),
                   alpha = alpha, lfcThreshold = 0, altHypothesis = "greaterAbs")
  
  f <- function(res){
    mcols(res, use.names = TRUE)
    sigtab <- res[which(res$padj < alpha), ]
  }
  
  f2 <- function(sigtab){
    cbind(as(sigtab, "data.frame"), as(tax_table(ps)[rownames(sigtab), ], "matrix"))
  }
  
  sigtab03   <- f(res03)
  sigtab09   <- f(res09)
  sigtab012  <- f(res012)
  sigtab39  <- f(res39)
  sigtab312  <- f(res312)
  sigtab912 <- f(res912)
  
  if (nrow(sigtab03) > 0){
    sigtab03  <- f2(sigtab03) 
    write.csv(sigtab03, file.path(path.out, "03.csv"))
  }
  if (nrow(sigtab09) > 0){
    sigtab09  <- f2(sigtab09) 
    write.csv(sigtab09, file.path(path.out, "09.csv"))
  }
  if (nrow(sigtab012) > 0){
    sigtab012 <- f2(sigtab012) 
    write.csv(sigtab012, file.path(path.out, "012.csv"))
  }
  if (nrow(sigtab39) > 0){
    sigtab39  <- f2(sigtab39) 
    write.csv(sigtab39, file.path(path.out, "39.csv"))
  }
  if (nrow(sigtab312) > 0){
    sigtab312 <- f2(sigtab312) 
    write.csv(sigtab312, file.path(path.out, "312.csv"))
  }
  if (nrow(sigtab912) > 0){
    sigtab912 <- f2(sigtab912) 
    write.csv(sigtab912, file.path(path.out, "912.csv"))
  }
  
}
```


## Define function for plotting 
```{r}
# Heat plot based on CheeseOutcome. 
ChouHeat <- function(ps, taxa, samType, w, h, path.out) {
  my_palette <- colorRampPalette(c("red", "yellow", "green"))(n = 299)
  # defines the color breaks manually for a "skewed" color transition
  col_breaks = c(seq(-1,0,length=100),      # for red
                 seq(0.01,0.8,length=100),  # for yellow
                 seq(0.81,1,length=100))    # for green

  ps <- ps %>% transform_sample_counts(function(x) x/sum(x) )  
  ps <- ps %>% tax_glom(taxrank = "Genus")
  
  taxa.df <- psmelt(ps)
  taxa.agg <- aggregate(Abundance ~ Genus + CheeseOutcome,
                        data = taxa.df,
                        mean)
  taxa.cast <- dcast(taxa.agg, Genus ~ CheeseOutcome, mean, value.var = "Abundance")
  # Make filter vector for the 14 selected cheddar taxa 
  TaxTab <- tax_table(ps) %>% as.data.frame()
  taxa_names(ps) <- TaxTab$Genus
  allTaxa <- taxa_names(ps)
  ps.notaxa <- prune_taxa(allTaxa[!(allTaxa %in% taxa)], ps)
  taxa.filt <- ps.notaxa@tax_table[,6] %>% as.character() #6 for genus level 
  # Filter
  taxa.cast[taxa.cast$Genus %in% taxa.filt,]$Genus <- "Other"
  taxa.cast.f <- taxa.cast[!(is.na(taxa.cast$Genus) | taxa.cast$Genus==""), ]
  
  # make as numeric matrix 
  row.names(taxa.cast.f) <- taxa.cast.f$Genus
  taxa.cast.f <- taxa.cast.f[,-1]
  
  
  png(path.out, width = w, height = h)
  superheat(taxa.cast.f,
            left.label.size = 0.4, 
            bottom.label.size = 0.1,
            order.rows = rev(order(rownames(taxa.cast.f))),
            #order.cols = c("1",2,3), # follow raw --> feed --> htst milk
            grid.hline = FALSE,
            title = samType,
            title.size = 6,
            title.alignment = "center") 
  dev.off()
}


# Heat plot based on MilkType
MilkHeat <- function(ps, taxa, samType, w, h, path.out) {
  my_palette <- colorRampPalette(c("red", "yellow", "green"))(n = 299)
  # defines the color breaks manually for a "skewed" color transition
  col_breaks = c(seq(-1,0,length=100),      # for red
                 seq(0.01,0.8,length=100),  # for yellow
                 seq(0.81,1,length=100))    # for green

  ps <- ps %>% transform_sample_counts(function(x) x/sum(x) )  
  ps <- ps %>% tax_glom(taxrank = "Genus")
  
  taxa.df <- psmelt(ps)
  taxa.agg <- aggregate(Abundance ~ Genus + SampleType,
                        data = taxa.df,
                        mean)
  taxa.cast <- dcast(taxa.agg, Genus ~ SampleType, mean, value.var = "Abundance")
  # Make filter vector for the 14 selected cheddar taxa 
  TaxTab <- tax_table(ps) %>% as.data.frame()
  taxa_names(ps) <- TaxTab$Genus
  allTaxa <- taxa_names(ps)
  ps.notaxa <- prune_taxa(allTaxa[!(allTaxa %in% taxa)], ps)
  taxa.filt <- ps.notaxa@tax_table[,6] %>% as.character() #6 for genus level 
  # Filter
  taxa.cast[taxa.cast$Genus %in% taxa.filt,]$Genus <- "Other"
  taxa.cast.f <- taxa.cast[!(is.na(taxa.cast$Genus) | taxa.cast$Genus==""), ]
  
  # make as numeric matrix 
  row.names(taxa.cast.f) <- taxa.cast.f$Genus
  taxa.cast.f <- taxa.cast.f[,-1]
  
  
  png(path.out, width = w, height = h)
  (superheat(taxa.cast.f,
            left.label.size = 0.4, 
            bottom.label.size = 0.1,
            order.rows = rev(order(rownames(taxa.cast.f))),
            #order.cols = c("1",2,3), # follow raw --> feed --> htst milk
            col.dendrogram = TRUE,
            grid.hline = FALSE,
            title = samType,
            title.size = 6,
            title.alignment = "center")) %>% print()
  dev.off()
}

# Heat plot based on TimeSinceCleaning. This function will plot only the top 30 taxa
TiSiHeat <- function(ps, samType, w, h, path.out) {
  my_palette <- colorRampPalette(c("red", "yellow", "green"))(n = 299)
  # defines the color breaks manually for a "skewed" color transition
  col_breaks = c(seq(-1,0,length=100),      # for red
                 seq(0.01,0.8,length=100),  # for yellow
                 seq(0.81,1,length=100))    # for green

  #ps <- ps %>% transform_sample_counts(function(x) x/sum(x) )  #get abundance in %
  ps <- ps %>% tax_glom(taxrank = "Genus")
  taxa.df <- psmelt(ps)
  taxa.agg <- aggregate(Abundance ~ Genus + TimeSinceCleaning,
                        data = taxa.df,
                        mean)
  taxa.cast <- dcast(taxa.agg, Genus ~ TimeSinceCleaning, mean, value.var = "Abundance")
  # need to change results from factor to numeric because of R
  row.names(taxa.cast) <- taxa.cast$Genus
  taxa.cast <- taxa.cast[, -1]
  indx <- sapply(taxa.cast, is.factor)
  taxa.cast[indx] <- lapply(taxa.cast[indx], function(x) as.numeric(as.character(x))) 
  taxa.cast30 <- cbind(taxa.cast, total = rowSums(taxa.cast)) #  need numeric values
  taxa.cast30$taxa <- rownames(taxa.cast30)
  taxa.cast30 <- head(arrange(taxa.cast30,desc(total)), n = 30)
  row.names(taxa.cast30) <- taxa.cast30$taxa
  taxa.cast30 <- taxa.cast30[, -c(5,6)]
  
  png(path.out, width = w, height = h)
  superheat(taxa.cast30,
            left.label.size = 0.4, 
            bottom.label.size = 0.1,
            order.rows = rev(order(rownames(taxa.cast30))),
            #order.cols = c("1",2,3), # follow raw --> feed --> htst milk
            grid.hline = FALSE,
            title = samType,
            title.size = 6,
            title.alignment = "center")
  dev.off()
}

# Time since cleaning heat plot 2 with all the taxa and taxa abudance clustering
TiSiHeat2 <- function(ps, samType, w, h, path.out) {
  my_palette <- colorRampPalette(c("red", "yellow", "green"))(n = 299)
  # defines the color breaks manually for a "skewed" color transition
  col_breaks = c(seq(-1,0,length=100),      # for red
                 seq(0.01,0.8,length=100),  # for yellow
                 seq(0.81,1,length=100))    # for green
  
  #ps <- ps %>% transform_sample_counts(function(x) x/sum(x) )  #get abundance in %
  ps <- ps %>% tax_glom(taxrank = "Genus")
  taxa.df <- psmelt(ps)
  taxa.agg <- aggregate(Abundance ~ Genus + TimeSinceCleaning,
                        data = taxa.df,
                        mean)
  taxa.cast <- dcast(taxa.agg, Genus ~ TimeSinceCleaning, mean, value.var = "Abundance")
  # need to change results from factor to numeric because of R
  row.names(taxa.cast) <- taxa.cast$Genus
  taxa.cast <- taxa.cast[, -1]
  indx <- sapply(taxa.cast, is.factor)
  taxa.cast[indx] <- lapply(taxa.cast[indx], function(x) as.numeric(as.character(x))) 
  taxa.cast30 <- cbind(taxa.cast, total = rowSums(taxa.cast)) #  need numeric values
  taxa.cast30$taxa <- rownames(taxa.cast30)
  taxa.cast30 <- head(arrange(taxa.cast30,desc(total)), n = 30)
  row.names(taxa.cast30) <- taxa.cast30$taxa
  taxa.cast30 <- taxa.cast30[, -c(5,6)]
  
  pdf(path.out, width = w, height = h)
  superheat(taxa.cast,
            left.label.size = 0.4, 
            bottom.label.size = 0.1,
            grid.hline = FALSE,
            title = samType,
            title.size = 6,
            scale = TRUE,
            row.dendrogram = TRUE,
            title.alignment = "center") %>% print()
  dev.off()
  
  # extract the clusters
  map <- superheat(taxa.cast, n.clusters.rows = 10)
  write.csv(map$membership.rows, file.path(paste(path.out,".csv", sep ='')))
}


# Stacked bar plot 
ChouBar <- function(ps, taxa, samType, w, h, path.out){
  ps <- ps %>% tax_glom(taxrank = "Genus") #%>%transform_sample_counts(function(x) x/sum(x) ) 
  TaxTab <- tax_table(ps) %>% as.data.frame()
  taxa_names(ps) <- TaxTab$Genus
  allTaxa <- taxa_names(ps)
  ps.notaxa <- prune_taxa(allTaxa[!(allTaxa %in% taxa)], ps)
  taxa.filt <- ps.notaxa@tax_table[,6] %>% as.character() #6 for genus level 
  
  # change the taxa that are not within the 13 common taxa to "Other" 
  TaxTab2 <- psmelt(ps)
  # merge/average per SampleType and CheeseOutcome
  TaxTab2_agg <- ddply(TaxTab2, c("SampleType", "CheeseOutcome", "Genus"),
                       summarise,
                       mean = mean(Abundance))
  TaxTab2_agg$Genus <- as.character(TaxTab2_agg$Genus)
  TaxTab2_agg[TaxTab2_agg$Genus %in% taxa.filt,]$Genus <- "Other"

  #Set colors for plotting
  mycol = c("#A6CEE3", "#438EC0", "#63A8A0", "#98D277", "#3BA432", "#F16667", 
            "#E62F27", "#F9A963", "#FE982C", "#ED8F47", "#C3AAD2", "#7D54A5", 
            "#B9A499", "#EAD27A", "#B15928")
  TaxTab2_agg$Genus = factor(TaxTab2_agg$Genus, levels = taxa)
  
  pdf(path.out, width = w, height = h)
  (ggplot(TaxTab2_agg, aes(x = CheeseOutcome, y = mean, fill = Genus)) +
    facet_wrap(~SampleType, scales = "fixed") +
    geom_bar(stat = "identity") + 
    scale_fill_manual(values = mycol)+
    # Remove x axis title
    theme(axis.title.x = element_blank(), text = element_text(size = 16)) + 
    guides(fill = guide_legend(reverse = FALSE, keywidth = 1, keyheight = 1)) +
    ylab("Log cells/ml")) %>% print() 
  dev.off()
}


# Stacked bar plot for time since cleaning
TiSinBar <- function(ps, taxa, samType, w, h, path.out){
  ps <- ps %>% tax_glom(taxrank = "Genus") #%>%transform_sample_counts(function(x) x/sum(x) ) 
  TaxTab <- tax_table(ps) %>% as.data.frame()
  taxa_names(ps) <- TaxTab$Genus
  allTaxa <- taxa_names(ps)
  ps.notaxa <- prune_taxa(allTaxa[!(allTaxa %in% taxa)], ps)
  taxa.filt <- ps.notaxa@tax_table[,6] %>% as.character() #6 for genus level 
  
  # change the taxa that are not within the 13 common taxa to "Other" 
  TaxTab2 <- psmelt(ps)
  # merge/average per SampleType and TimeSinceCleaning
  TaxTab2_agg <- aggregate(Abundance ~ SampleType + TimeSinceCleaning + Genus,
                           data = TaxTab2,
                           mean)
  TaxTab2_agg$Genus <- as.character(TaxTab2_agg$Genus)
  TaxTab2_agg[TaxTab2_agg$Genus %in% taxa.filt,]$Genus <- "Other"

  #Set colors for plotting
  mycol = c("#A6CEE3", "#438EC0", "#63A8A0", "#98D277", "#3BA432", "#F16667", 
            "#E62F27", "#F9A963", "#FE982C", "#ED8F47", "#C3AAD2", "#7D54A5", 
            "#B9A499", "#EAD27A", "#B15928")
  TaxTab2_agg$Genus = factor(TaxTab2_agg$Genus, levels = taxa)
  
  pdf(path.out, width = w, height = h)
  print(ggplot(TaxTab2_agg, aes(x = TimeSinceCleaning, y = Abundance, fill = Genus)) +
                facet_wrap(~SampleType, scales = "fixed") +
                geom_bar(stat = "identity") + 
                scale_fill_manual(values = mycol)+
                # Remove x axis title
                theme_bw()+
                theme(text = element_text(size = 15)) + 
                guides(fill = guide_legend(reverse = FALSE, keywidth = 1, keyheight = 1)) +
                ylab("Log cells/ml")) 
  dev.off()
}
```


## Prepare phyloseq objects 
```{r}
# remake a unrarefied ps object with low coverage samples filtered
## The starting ps is removed of ASV with undetermined phylum and ASV < 0.00005
ps05.Phy05 <- readRDS(file.path(path, "decontam/phyloseq/ps05.Phy05.rds"))
ps05.Phy05
# add tree
tree <- read_tree(file.path(path, "decontam/phyloseq/tree.nwk"))
ps05.Phy05.tre <- merge_phyloseq(ps05.Phy05, phy_tree(tree))
saveRDS(ps05.Phy05.tre, file.path(path, "decontam/phyloseq/ps05.Phy05.tre.rds"))
ps05.Phy05.tre

# remove sample < 100 reads
ps100 <- prune_samples(sample_sums(ps05.Phy05.tre) >= 100, ps05.Phy05.tre)
saveRDS(ps100, file.path(path, "decontam/phyloseq/ps100.rds"))
ps100

# remove all NC and seq control samples 
ps100.sam <- subset_samples(ps100, SampleOrCtrl == "Sample" )
#ps100.sam <- ps100.sam %>% tax_glom(taxrank = "Genus")
saveRDS(ps100.sam, file.path(path, "decontam/phyloseq/ps100.sam.rds"))
ps100.sam

# All milk samples
ps.allmilk <- subset_samples(ps100.sam, SampleType %in% c("HTST_milk","HTST_feed", "Raw_milk"))

# Feed and HTST milk with information on TimeSinceCleaning
ps.milk <- subset_samples(ps100.sam, SampleType %in% c("HTST_milk","HTST_feed"))
ps.milk
## Read in new sample information dataframe 
sam.milk <- read.csv(file.path(path, "mapping/sam_milk_HTST0369.csv"))
rownames(sam.milk) <- sam.milk$SampleID
## Merge as a new phyloseq object
tree <- phy_tree(ps.milk)
tax <- tax_table(ps.milk)
otu <- otu_table(ps.milk)
ps.milk <- phyloseq(otu_table(otu, taxa_are_rows = FALSE), tax_table(tax),sample_data(sam.milk), phy_tree(tree))
saveRDS(ps.milk, file.path(path, "milk/ps.milk.rds"))
ps.milk

# Feed and HTST milk that are paired with cheddar (based on the 4hr timeline)
samdf.4hr <- read.csv(file.path(path, "mapping/Run1-5_samdf-mc2_4hrTL-NoExtremeMilk.csv"))
rownames(samdf.4hr) <- samdf.4hr$SampleID
ps.milk4 <- phyloseq(otu_table(otu, taxa_are_rows = FALSE), tax_table(tax),sample_data(samdf.4hr), phy_tree(tree))
saveRDS(ps.milk4, file.path(path, "milk/ps.milk4.rds"))
ps.milk4
```


## Taxonomic DA of Cheddar 
```{r}
# 1 # All Cheddar samples 
ps.ched <- subset_samples(ps100.sam, CheeseType %in% "Cheddar")
ps.ched <- subset_samples(ps.ched, 
                          SampleType %in% c("Cheese_0D","Cheese_30D","Cheese_90D","Cheese_120D"))
saveRDS(ps.ched, file.path(path, "cheddar/ps.ched.rds"))
ps.ched
# PMA 
ps.chedY <- subset_samples(ps.ched, PMA %in% "Y")
saveRDS(ps.chedY, file.path(path, "cheddar/ps.chedY.rds"))
ps.chedY
CheeseOutcomeDAnoMean(ps.chedY, path.out = file.path(path, "cheddar/deseq_table/sigtab.chedY.csv"))
# PMA untreated 
ps.chedN <- subset_samples(ps.ched, PMA %in% "N")
saveRDS(ps.chedN, file.path(path, "cheddar/ps.chedN.rds"))
ps.chedN
CheeseOutcomeDAnoMean(ps.chedN, path.out = file.path(path, "cheddar/deseq_table/sigtab.chedN.csv"))


# 2 # All Cheddar samples without Lactococcus
# PMA treated 
ps.chedY.noL <- subset_taxa(ps.chedY, !Genus %in% "Lactococcus")
saveRDS(ps.chedY.noL, file.path(path, "cheddar/ps.chedY.noL.rds"))
ps.chedY.noL
CheeseOutcomeDA(ps.chedY.noL, 
                path.out = file.path(path, "cheddar/deseq_table/sigtab.chedY.noL.csv"))
ChedDA(ps.chedY.noL, path.out = file.path(path, "cheddar/deseq_table_chedD/chedY"))
# PMA untreated 
ps.chedN.noL <- subset_taxa(ps.chedN, !Genus %in% "Lactococcus")
saveRDS(ps.chedN.noL, file.path(path, "cheddar/ps.chedN.noL.rds"))
ps.chedN.noL
CheeseOutcomeDA(ps.chedN.noL, 
                path.out = file.path(path, "cheddar/deseq_table/sigtab.chedN.noL.csv"))
ChedDA(ps.chedN.noL, path.out = file.path(path, "cheddar/deseq_table_chedD/chedN"))


# 3 # 0-day Cheddar samples without Lactococcus
ps.chedY.noL0 <- subset_samples(ps.chedY.noL, SampleType %in% "Cheese_0D")
saveRDS(ps.chedY.noL0, file.path(path, "cheddar/ps.chedY.noL0.rds"))
ps.chedY.noL0
CheeseOutcomeDA(ps.chedY.noL0, 
                path.out = file.path(path, "cheddar/deseq_table/sigtab.chedY.noL0.csv"))

# 4 # 30-day Cheddar samples without Lactococcus
ps.chedY.noL30 <- subset_samples(ps.chedY.noL, SampleType %in% "Cheese_30D")
saveRDS(ps.chedY.noL30, file.path(path, "cheddar/ps.chedY.noL30.rds"))
ps.chedY.noL30
CheeseOutcomeDAnoMean(ps.chedY.noL30, 
                path.out = file.path(path, "cheddar/deseq_table/sigtab.chedY.noL30.csv"))

# 5 # 90-day Cheddar samples without Lactococcus
ps.chedY.noL90 <- subset_samples(ps.chedY.noL, SampleType %in% "Cheese_90D")
saveRDS(ps.chedY.noL90, file.path(path, "cheddar/ps.chedY.noL90.rds"))
ps.chedY.noL90
CheeseOutcomeDAnoMean(ps.chedY.noL90, 
                path.out = file.path(path, "cheddar/deseq_table/sigtab.chedY.noL90.csv"))

# 6 # 120-day Cheddar samples without Lactococcus
ps.chedY.noL120 <- subset_samples(ps.chedY.noL, SampleType %in% "Cheese_120D")
saveRDS(ps.chedY.noL120, file.path(path, "cheddar/ps.chedY.noL120.rds"))
ps.chedY.noL120
CheeseOutcomeDAnoMean(ps.chedY.noL120, 
                path.out = file.path(path, "cheddar/deseq_table/sigtab.chedY.noL120.csv"))
```


## Taxonomic DA of HTST feed and milk that has not been adjusted for cell count using qPCR results
This is PMA treated samples
```{r}
# 7 # HTST milk 
ps.htst <- subset_samples(ps.milk, SampleType %in% "HTST_milk")
saveRDS(ps.htst, file.path(path, "milk/ps.htst.rds"))
ps.htst
ps.htstY <- subset_samples(ps.htst, PMA %in% "Y")
saveRDS(ps.htstY, file.path(path, "milk/ps.htstY.rds"))
CheeseOutcomeDA(ps.htstY, 
                path.out = file.path(path, "milk/deseq_table/sigtab.htstY.csv"))
ChouHeat(ps.htstY, samType = "PMA treated HTST milk", w = 8.5, h = 11,
         path.out = file.path(path, "milk/fig/htstY.pdf"))
TiSinCleaningDA(ps.htstY, 
                path.out = file.path(path, "milk/deseq_table_TiSinCleaning/htstY"))
TiSiHeat(ps.htstY, samType = "PMA treated HTST milk", w = 8.5, h = 11,
         path.out = file.path(path, "milk/fig_TiSinCleaning/htstY.pdf"))
TiSiHeat2(ps.htstY, samType = "PMA treated HTST milk", w = 8.5, h = 20,
         path.out = file.path(path, "milk/fig_TiSinCleaning/htstY2.pdf"))
# PMA untreated
ps.htstN <- subset_samples(ps.htst, PMA %in% "N")
saveRDS(ps.htstN, file.path(path, "milk/ps.htstN.rds"))
CheeseOutcomeDA(ps.htstN, 
                path.out = file.path(path, "milk/deseq_table/sigtab.htstN.csv"))
TiSiHeat2(ps.htstN, samType = "PMA untreated HTST milk", w = 8.5, h = 20,
          path.out = file.path(path, "milk/fig_TiSinCleaning/htstN2.pdf"))

# 13 # HTST feed 
ps.feed <- subset_samples(ps.milk, SampleType %in% "HTST_feed")
ps.feedY <- subset_samples(ps.feed, PMA %in% "Y")
CheeseOutcomeDAnoMean(ps.feedY, 
                path.out = file.path(path, "milk/deseq_table/sigtab.feedY.csv"))
TiSinCleaningDA(ps.feedY, 
                path.out = file.path(path, "milk/deseq_table_TiSinCleaning/feedY"))
# PMA untreated
ps.feedN <- subset_samples(ps.feed, PMA %in% "N")
CheeseOutcomeDAnoMean(ps.feedN, 
                path.out = file.path(path, "milk/deseq_table/sigtab.feedN.csv"))


# 8 # HTST milk that are paired with cheese (4hr timeline)
ps.htst4 <- subset_samples(ps.milk4, SampleType %in% "HTST_milk")
saveRDS(ps.htst4, file.path(path, "milk/ps.htst4.rds"))
ps.htst4
ps.htst4Y <- subset_samples(ps.htst4, PMA %in% "Y")
saveRDS(ps.htstY, file.path(path, "milk/ps.htstY.rds"))
ps.htst4Y
CheeseOutcomeDA(ps.htst4Y, 
                path.out = file.path(path, "milk/deseq_table/sigtab.htst4Y.csv"))
ChouHeat(ps.htst4Y, comm14,
         samType = "PMA treated paired HTST milk", w = 450, h = 700,
         path.out = file.path(path, "milk/fig/htst4Y.png"))


# 12 # HTST feed that are sampled around the 4hr timeline
ps.feed4 <- subset_samples(ps.milk4, SampleType %in% "HTST_feed")
saveRDS(ps.feed4, file.path(path, "milk/ps.feed4.rds"))
ps.feed4
ps.feed4Y <- subset_samples(ps.feed4, PMA %in% "Y")
saveRDS(ps.feed4Y, file.path(path, "milk/ps.feed4Y.rds"))
ps.feed4Y
CheeseOutcomeDA(ps.feed4Y, 
                path.out = file.path(path, "milk/deseq_table/sigtab.feed4Y.csv"))
ChouHeat(ps.feed4Y, comm14,
         samType = "PMA treated paired HTST feed", w = 450, h = 700,
         path.out = file.path(path, "milk/fig/feed4Y.png"))
```


## Taxonomic DA of HTST milk that ARE adjusted for cell count using qPCR results
DESeq2 keeps giving error because after qpcr adjustment, it is not raw counts anymore.
```{r eval=FALSE, include=FALSE}
qPCR <- read.csv(file.path(path, "mapping/Milk_sample_qPCR.csv"))
# transform to percentage
ps.milkq <- ps.milk %>% 
  tax_glom(taxrank = "Genus") %>% 
  transform_sample_counts(function(x) x/sum(x) )  
otuq <- otu_table(ps.milkq) %>% data.frame()
otuq$Sampleid <- row.names(otuq)
colnames(otuq) <- gsub("X","",colnames(otuq))
otu.q <- merge(x = qPCR, y = otuq, by.x = "SampleID", by.y = "Sampleid")
coln <- colnames(otu.q[,c(20:91)]) #  character vector with all feature names
otu.q2 <- otu.q %>% mutate_at(.vars =  coln, .funs = funs(.*LogCellsML))
row.names(otu.q2) <- otu.q2[,1] #sample id as row names for formatting with phyloseq
otu.q3 <- otu.q2[,-c(1:19)] #remove all the meta information so the resulting otu table can be transformed into a numeric matrix
otu.q3 <- as.matrix(otu.q3)

# Merge as a new phyloseq object
ps.milkq <- phyloseq(otu_table(otu.q3, taxa_are_rows = FALSE), 
                     tax_table(ps.milkq),
                     sample_data(ps.milkq), 
                     phy_tree(ps.milkq))
# remove features that sum up to 0
#ps.milkq <- filter_taxa(ps.milkq, function(x) sum(x) > 0, TRUE)
#ps.milkq <- prune_samples(sample_sums(ps.milkq)>=0, ps.milkq)
saveRDS(ps.milkq, file.path(path, "milk/ps.milkq.rds"))
ps.milkq


# 9 # HTST milk that has been adjusted with qPCR values
ps.htstq <- subset_samples(ps.milkq, SampleType %in% "HTST_milk")
saveRDS(ps.htstq, file.path(path, "milk/ps.htstq.rds"))
ps.htstqY <- subset_samples(ps.htstq, PMA %in% "Y")
saveRDS(ps.htstqY, file.path(path, "milk/ps.htstqY.rds"))
ps.htstqN <- subset_samples(ps.htstq, PMA %in% "N")
saveRDS(ps.htstqN, file.path(path, "milk/ps.htstqN.rds"))
CheeseOutcomeDA(ps.htstqY, 
                path.out = file.path(path, "milk/deseq_table/sigtab.htstqY.csv"))
ChouBar(ps.htstqY, 
        taxa = comm14,
        samType = "PMA treated HTST milk", w = 6, h = 5,
        path.out = file.path(path, "milk/fig/htstqY2.pdf"))
# Has error -->  TiSinCleaningDAnoMean(ps.htstqY, path.out = file.path(path, "milk/deseq_table_TiSinCleaning/htstqY"))


# 14 # HTST feed that has been adjusted with qPCR values
ps.feedq <- subset_samples(ps.milkq, SampleType %in% "HTST_feed")
ps.feedqY <- subset_samples(ps.feedq, PMA %in% "Y")
# Has error -->  TiSinCleaningDA(ps.feedqY, path.out = file.path(path, "milk/deseq_table_TiSinCleaning/feedqY"))


# 10 # HTST milk around the 4hr timeline that has been adjusted with qPCR values
ps.milkq4 <- phyloseq(otu_table(ps.milkq), tax_table(ps.milkq),sample_data(samdf.4hr), phy_tree(ps.milkq))
ps.milkq4Y <- subset_samples(ps.milkq4, PMA %in% "Y")
ps.milkq4N <- subset_samples(ps.milkq4, PMA %in% "N")
ps.htstq4Y <- subset_samples(ps.milkq4Y, SampleType %in% "HTST_milk")
# Has error -->  CheeseOutcomeDAnoMean(ps.htstq4Y, path.out = file.path(path, "milk/deseq_table/sigtab.htstq4Y.csv"))
ChouHeat(ps.htstq4Y, comm18,
         samType = "PMA treated HTST milk", w = 450, h = 700,
         path.out = file.path(path, "milk/fig/htstq4Y.png"))

# 13 # HTST feed around the 4hr timeline that has been adjusted with qPCR values
ps.feedq4Y <- subset_samples(ps.milkq4Y, SampleType %in% "HTST_feed")
CheeseOutcomeDA(ps.feedq4Y, path.out = file.path(path, "milk/deseq_table/sigtab.feedq4Y.csv"))
ChouHeat(ps.feedq4Y, comm18,
         samType = "PMA treated HTST feed", w = 450, h = 700,
         path.out = file.path(path, "milk/fig/feedq4Y.png"))

# 14 # Both HTST feed and HTST milk around the 4hr timeline that has been adjusted with qPCR
ChouBar(ps.milkq4Y, 
        taxa = comm14,
        samType = "PMA treated HTST milk", w = 6, h = 5,
        path.out = file.path(path, "milk/fig/milkq4Y.pdf"))
ChouBar(ps.milkq4N, 
        taxa = comm14,
        samType = "PMA untreated HTST milk", w = 6, h = 5,
        path.out = file.path(path, "milk/fig/milkq4N.pdf"))

```


## Taxanomic DA based on milk SampleType
```{r}
# 11 # taxa DA based on milk type 
ps.allmilkY <- subset_samples(ps.allmilk, PMA %in% "Y")
MilkDA(ps.allmilkY, 
       path.out = file.path(path, "milk/deseq_table_SampleType/milkY"))

ps.allmilkN <- subset_samples(ps.allmilk, PMA %in% "N")
MilkDA(ps.allmilkN, 
       path.out = file.path(path, "milk/deseq_table_SampleType/milkN"))
```












## Line figures for each taxa 
```{r}
# Define function for line plot
LinePlot <- function(ps, x, w, h, path.out){
  melt <- psmelt(ps) 
  agg <- ddply(melt, c("TimeSinceCleaning","SampleType","PMA", "Genus"),
               summarise,
               N = length(Abundance),
               mean = mean(Abundance),
               sd = sd(Abundance),
               se = sd / sqrt(N))
  
  pdf(path.out, width = w, height = h)
  print(ggplot(subset(agg, Genus %in% x), aes(x = TimeSinceCleaning, y = mean, group = PMA,color = PMA)) +
               geom_line(size = 0.9)+
               geom_point(size = 1)+
               geom_errorbar(aes(ymin = mean-se, ymax = mean+se),width = 0.1, position=position_dodge(0.05))+
               facet_wrap(~ SampleType, scales = "fixed")+
               ylab(x) +
               theme(text = element_text(size = 15)))
  dev.off()
}

# Define function for line plot 2 that only plots PMA treated only
LinePlot2 <- function(ps, x, w, h, path.out){
  melt <- psmelt(ps) 
  agg <- ddply(melt, c("TimeSinceCleaning","SampleType","PMA", "Genus"),
               summarise,
               N = length(Abundance),
               mean = mean(Abundance),
               sd = sd(Abundance),
               se = sd / sqrt(N))
  
  pdf(path.out, width = w, height = h)
  print(ggplot(subset(agg, Genus %in% x), aes(x = TimeSinceCleaning, y = mean, color = SampleType)) +
              geom_line(size = 0.9)+
              geom_point(size = 1)+
              geom_errorbar(aes(ymin = mean-se, ymax = mean+se),width = 0.1, position=position_dodge(0.05))+
              ylab(x) +
              theme_bw() +
              theme(text = element_text(size = 15)))
  dev.off()
}

# plot total load
qPCR.agg <- ddply(qPCR, c("TimeSinceCleaning","Description","PMA"),
                  summarise,
                  N = length(LogCellsML),
                  mean = mean(LogCellsML),
                  sd = sd(LogCellsML),
                  se = sd / sqrt(N))
# subset to include only timepoints that have N>1
qPCR.agg <- qPCR.agg[qPCR.agg$N != '1',]

pdf(file.path(path, "milk/fig_TiSinCleaning/total.pdf"), w = 4.5, h = 3.5)
ggplot(qPCR.agg, aes(x = TimeSinceCleaning, y = mean, group = PMA,color = PMA)) +
  geom_line(size = 0.9)+
  geom_point(size = 1)+
  geom_errorbar(aes(ymin = mean-se, ymax = mean+se),width = 0.1, position=position_dodge(0.05))+
  facet_wrap(~ Description, scales = "fixed")+
  ylab("Log cells/ml") +
  theme(text = element_text(size = 15))
dev.off()
# PMA treated
pdf(file.path(path, "milk/fig_TiSinCleaning/Y/total.pdf"), w = 4.5, h = 3.5)
ggplot(subset(qPCR.agg, PMA == "Y"), aes(x = TimeSinceCleaning, y = mean, color = Description)) +
  geom_line(size = 0.9)+
  geom_point(size = 1)+
  geom_errorbar(aes(ymin = mean-se, ymax = mean+se),width = 0.1, position=position_dodge(0.05))+
  ylab("Log cells/ml") +
  theme_bw() +
  theme(text = element_text(size = 15)) 
dev.off()
# PMA UNtreated
pdf(file.path(path, "milk/fig_TiSinCleaning/N/total.pdf"), w = 4.5, h = 3.5)
ggplot(subset(qPCR.agg, PMA == "N"), aes(x = TimeSinceCleaning, y = mean, color = Description)) +
  geom_line(size = 0.9)+
  geom_point(size = 1)+
  geom_errorbar(aes(ymin = mean-se, ymax = mean+se),width = 0.1, position=position_dodge(0.05))+
  ylab("Log cells/ml") +
  theme_bw() +
  theme(text = element_text(size = 15)) 
dev.off()



# loop for all specific taxa
for(i in 1:length(tax_table(ps.milkq)[,6])){
  LinePlot(ps.milkq, x = tax_table(ps.milkq)[,6][i], w = 4.5, h = 3.5, 
           path.out = file.path(path, "milk/fig_TiSinCleaning", paste(tax_table(ps.milkq)[,6][i],".pdf", sep ='')))
}
# Loop for all sepcific taxa ## PMA treated
for(i in 1:length(tax_table(ps.milkqY)[,6])){
  LinePlot2(ps.milkqY, x = tax_table(ps.milkqY)[,6][i], w = 4.5, h = 3.5, 
           path.out = file.path(path, "milk/fig_TiSinCleaning/Y", paste(tax_table(ps.milkqY)[,6][i],".pdf", sep ='')))
}
# Loop for all sepcific taxa ## PMA UNtreated
for(i in 1:length(tax_table(ps.milkqN)[,6])){
  LinePlot2(ps.milkqN, x = tax_table(ps.milkqN)[,6][i], w = 4.5, h = 3.5, 
            path.out = file.path(path, "milk/fig_TiSinCleaning/N", paste(tax_table(ps.milkqY)[,6][i],".pdf", sep ='')))
}
# Loop for all sepcific taxa in HTST milk ## PMA treated
for(i in 1:length(tax_table(ps.htstqY)[,6])){
  LinePlot2(ps.htstqY, x = tax_table(ps.htstqY)[,6][i], w = 4.5, h = 3.5, 
            path.out = file.path(path, "milk/fig_TiSinCleaning/htstqY", paste(tax_table(ps.htstqY)[,6][i],".pdf", sep ='')))
}
# percent looks the same as the one adjusted with qPCR
#ps.htstqY.per <- ps.htstqY %>% transform_sample_counts(function(x) x/sum(x) )
#for(i in 1:length(tax_table(ps.htstqY.per)[,6])){
#  LinePlot2(ps.htstqY.per, x = tax_table(ps.htstqY.per)[,6][i], w = 4.5, h = 3.5, 
#            path.out = file.path(path, "milk/fig_TiSinCleaning/htstY", paste(tax_table(ps.htstqY.per)[,6][i],".pdf", sep ='')))
#}
# Loop for all sepcific taxa in HTST milk ## PMA UNtreated
for(i in 1:length(tax_table(ps.htstqN)[,6])){
  LinePlot2(ps.htstqN, x = tax_table(ps.htstqN)[,6][i], w = 4.5, h = 3.5, 
            path.out = file.path(path, "milk/fig_TiSinCleaning/htstqN", paste(tax_table(ps.htstqN)[,6][i],".pdf", sep ='')))
}

# Aggregate some plots on to 1 figure
LinePlot3 <- function(ps, x, w, h, path.out){
  melt <- psmelt(ps) 
  agg <- ddply(melt, c("TimeSinceCleaning","SampleType","PMA", "Genus"),
               summarise,
               N = length(Abundance),
               mean = mean(Abundance),
               sd = sd(Abundance),
               se = sd / sqrt(N))
  
  pdf(path.out, width = w, height = h)
  print(ggplot(subset(agg, Genus %in% x), aes(x = TimeSinceCleaning, y = mean, color = Genus)) +
          geom_line(size = 0.9)+
          geom_point(size = 1)+
          geom_errorbar(aes(ymin = mean-se, ymax = mean+se),width = 0.1, position=position_dodge(0.05))+
          ylab("Log cells/ml") +
          theme_bw() +
          theme(text = element_text(size = 15)))
  dev.off()
}

# taxa that are resistent to cleaning and significant 
res.hi <- c("Staphylococcus", "Turicibacter")  #  high abundance taxa, > 0.5
res.med <- c("Acinetobacter", "Bacillus", "Lactococcus") #  median abundance 0.1-0.5
res.lo <- c("Epulopiscium", "Lactobacillus", "Macrococcus", "Mycoplasma") #  0.02- 0.1
res.xs <- c("Bacteroides", "Bifidobacterium", "Brevibacillus", "Enhydrobacter", 
            "Enterococcus", "Escherichia")   #  < 0.02

# HTST milk PMA treated 
LinePlot3(ps.htstqY, x = res.hi, w = 4.5, h = 3.5, 
          path.out = file.path(path, "milk/fig_TiSinCleaning/htstqY/hi_taxa.pdf"))
LinePlot3(ps.htstqY, x = res.med, w = 4.5, h = 3.5, 
          path.out = file.path(path, "milk/fig_TiSinCleaning/htstqY/med_taxa.pdf"))
LinePlot3(ps.htstqY, x = res.lo, w = 4.5, h = 3.5, 
          path.out = file.path(path, "milk/fig_TiSinCleaning/htstqY/lo_taxa.pdf"))
LinePlot3(ps.htstqY, x = res.xs, w = 4.5, h = 3.5, 
          path.out = file.path(path, "milk/fig_TiSinCleaning/htstqY/xs_taxa.pdf"))
```

## ANOVA line figures for each taxa 
```{r}
TiSiCleanANOVA <- function(ps, x, path.out){
  agg <- psmelt(ps)
  
  fit <- aov(Abundance ~ TimeSinceCleaning, data = subset(agg, Genus %in% x))
  y <- summary(fit)[[1]]
  # if overall value < 0.05, move on with post hoc test
  if (y$`Pr(>F)`[1] == "NaN"){  # if might be NaN value if taxa abudance is 0
    print(paste(x, "abundance is 0", sep = ' '))
  } else if (y$`Pr(>F)`[1] < 0.08){
    attach(subset(agg, Genus %in% x))
    z <- pairwise.t.test(Abundance, TimeSinceCleaning, p.adjust.method = "none")[[3]]
    detach()
    write.csv(z, path.out)
  }
}

# Loop ANOVA test for all taxa in HTST milk ## PMA treated
for(i in 1:length(tax_table(ps.htstqY)[,6])){
  TiSiCleanANOVA(ps.htstqY, x = tax_table(ps.htstqY)[,6][i],
                 path.out = file.path(path, "milk/fig_TiSinCleaning/htstqY/ANOVA", paste(tax_table(ps.htstqY)[,6][i],".csv", sep ='')))
}
# Loop ANOVA test for all taxa in HTST milk ## PMA UNtreated
for(i in 1:length(tax_table(ps.htstqN)[,6])){
  TiSiCleanANOVA(ps.htstqN, x = tax_table(ps.htstqN)[,6][i],
                 path.out = file.path(path, "milk/fig_TiSinCleaning/htstqN/ANOVA", paste(tax_table(ps.htstqN)[,6][i],".csv", sep ='')))
}
# Loop ANOVA test for all taxa in HTST feed ## PMA treated
for(i in 1:length(tax_table(ps.feedqY)[,6])){
  TiSiCleanANOVA(ps.feedqY, x = tax_table(ps.feedqY)[,6][i],
                 path.out = file.path(path, "milk/fig_TiSinCleaning/feedqY/ANOVA", paste(tax_table(ps.feedqY)[,6][i],".csv", sep ='')))
}

# ANOVA function for qPCR files
qPCRanova <- function(x, path.out){
  x <- x[x$TimeSinceCleaning %in% c("0","3","6","9"),]
  
  aov(LogCellsML ~ TimeSinceCleaning, data = x) %>% summary()

  attach(x)
  z <- pairwise.t.test(LogCellsML, TimeSinceCleaning, p.adjust.method = "none")[[3]]
  detach()
  
  write.csv(z, path.out)
}

# ANOVA for total cell count in HTST milk ## PMA treated
qPCRY <- subset(qPCR, PMA == "Y")
qPCR.htstY <- subset(qPCRY, Description == "HTST milk")
qPCRanova(qPCR.htstY, path.out = file.path(path, "milk/fig_TiSinCleaning/htstqY/ANOVA/total.csv"))
# ANOVA for total cell count in HTST feed ## PMA treated
qPCR.feedY <- subset(qPCRY, Description == "HTST feed")
qPCRanova(qPCR.feedY, path.out = file.path(path, "milk/fig_TiSinCleaning/feedqY/ANOVA/total.csv"))
```


## Some additional Plotting and saving as pdf
```{r}
ChouHeat(ps.chedY.noL0, ched14, 
         samType = "PMA treated D0 Cheddar", w = 450, h = 700,
         path.out = file.path(path, "cheddar/fig/chedY.noL0.png"))
ChouHeat(ps.chedY.noL30, ched14,
         samType = "PMA treated D30 Cheddar", w = 450, h = 700,
         path.out = file.path(path, "cheddar/fig/chedY.noL30.png"))
ChouHeat(ps.chedY.noL90, ched14, 
         samType = "PMA treated D90 Cheddar", w = 450, h = 700,
         path.out = file.path(path, "cheddar/fig/chedY.noL90.png"))
ChouHeat(ps.chedY.noL120, ched14,
         samType = "PMA treated D120 Cheddar", w = 450, h = 700,
         path.out = file.path(path, "cheddar/fig/chedY.noL120.png"))


MilkHeat(ps.allmilkY, comm14, 
         samType = "PMA treated milk", w = 450, h = 700,
         path.out = file.path(path, "milk/fig_SampleType/milkY.png"))
MilkHeat(ps.allmilkN, comm14, 
         samType = "PMA untreated milk", w = 450, h = 700,
         path.out = file.path(path, "milk/fig_SampleType/milkN.png"))
MilkHeat2(ps.allmilkY, 
          samType = "PMA treated milk", w = 15, h = 20,
          path.out = file.path(path, "milk/fig_SampleType/milkY2.pdf"))
MilkHeat2(ps.allmilkN, 
          samType = "PMA untreated milk", w = 15, h = 20,
          path.out = file.path(path, "milk/fig_SampleType/milkN2.pdf"))


ps.milkY <- subset_samples(ps.milk, PMA %in% "Y")
ChouBar(ps.milkY, 
        taxa = comm14,
        samType = "PMA treated HTST milk", w = 6, h = 5,
        path.out = file.path(path, "milk/fig/milkY.pdf"))


ps.milkqY <- subset_samples(ps.milkq, PMA %in% "Y")
ChouBar(ps.milkqY, 
        taxa = comm14,
        samType = "PMA treated HTST milk", w = 6, h = 5,
        path.out = file.path(path, "milk/fig/milkqY.pdf"))
TiSinBar(ps.milkqY, 
        taxa = comm14,
        samType = "PMA treated HTST milk", w = 6, h = 5,
        path.out = file.path(path, "milk/fig_TiSinCleaning/htstqY2.pdf"))
ps.milkqN <- subset_samples(ps.milkq, PMA %in% "N")
TiSinBar(ps.milkqN, 
         taxa = comm14,
         samType = "PMA untreated HTST milk", w = 6, h = 5,
         path.out = file.path(path, "milk/fig_TiSinCleaning/htstqN2.pdf"))

```

