---
title: "taxa_DESeq2_phyloseq"
author: "Zeya Zhengyao Xue"
date: "October 8, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


I used old .Rmd and html with rarefied feature table to generate alpha and beta 
diversity figures. This script is focused on differential abudance testing on 
taxonomy profile.

## Setting up
```{r}
library(phyloseq);packageVersion("phyloseq")
library(DESeq2);packageVersion("DESeq2")
library(ggplot2)
library(reshape2)
library(superheat)
#library(cowplot)
library(plyr)
library(dplyr)
#library(RColorBrewer)
#library(microbiomeSeq)

# Define project path
path <- "G:/My Drive/UC_Davis/Marco_lab/milk_microbiota/Hilmar_weekly_samples/QIIME2/"
# copy the .Rmd file to the path directory to generate html report

set.seed(123) #  for reproducibility

# character vector for the comm 14 most abundant taxa
comm14 <- c("Acinetobacter", "Bacillus", "Corynebacterium", "Clostridium",
                 "Enterococcus", "Escherichia", "Lactococcus", "Macrococcus",
                 "Mycoplasma", "Pseudomonas", "Staphylococcus", "Streptococcus",
                 "Thermus", "Turicibacter", "Other")
```


## Define function to use DESeq2 for differential analysis 
```{r}
# Cheese Outcome Differential abundance and generate diagnostic plots and write sig tables 
## fold change is slits vs no slits
## Uses default Benjamini-Hochberg pvalue adjust
CheeseOutcomeDA <- function(ps, path.out) {
  ps <- ps %>% tax_glom(taxrank = "Genus")
  psdds <- phyloseq_to_deseq2(ps, ~ CheeseOutcome)
  
  dds <- DESeq(psdds, test = "Wald", fitType = "local") 
  plotDispEsts(dds, ylim = c(1e-6, 1e2)) 
  alpha <- 0.5 #  padj for indepenent filtering and expect FDR < alpha
  res <- results(dds, alpha = alpha, 
                 lfcThreshold = 0, 
                 altHypothesis = "greaterAbs")
  mcols(res, use.names=TRUE)  
  plotMA(res) # diagnostic description
  hist(res$pvalue, breaks=20, col="grey" ) # diagnostic plot
  sigtab = res[which(res$padj < alpha), ] 
  if (nrow(sigtab) == 0) {
    print("There are no significant features whose padj < 0.5 and |lfc| > 0")
  } else {
    sigtab <- cbind(as(sigtab, "data.frame"), as(tax_table(ps)[rownames(sigtab), ], "matrix"))
    write.csv(sigtab, file.path(path.out))
  }
}


# TimeSinceCleaning Differential abundance and generate diagnostic plots and write sig tables 
TiSinCleaningDA <- function(ps, path.out) {
  ps <- ps %>% tax_glom(taxrank = "Genus")
  sample_data(ps)$TimeSinceCleaning <- factor(sample_data(ps)$TimeSinceCleaning)
  psdds <- phyloseq_to_deseq2(ps, design = ~ TimeSinceCleaning)
  
  dds <- DESeq(psdds, test = "Wald", fitType = "local") 
  plotDispEsts(dds, ylim = c(1e-6, 1e2)) 
  alpha <- 0.5 #  padj for indepenent filtering and expect FDR < alpha

  res03 <- results(dds, contrast = c("TimeSinceCleaning", "0", "3"),
                   alpha = alpha, lfcThreshold = 0, altHypothesis = "greaterAbs")
  res06 <- results(dds, contrast = c("TimeSinceCleaning", "0", "6"),
                   alpha = alpha, lfcThreshold = 0, altHypothesis = "greaterAbs")
  res09 <- results(dds, contrast = c("TimeSinceCleaning", "0", "9"),
                   alpha = alpha, lfcThreshold = 0, altHypothesis = "greaterAbs")
  res36 <- results(dds, contrast = c("TimeSinceCleaning", "3", "6"),
                   alpha = alpha, lfcThreshold = 0, altHypothesis = "greaterAbs")
  res39 <- results(dds, contrast = c("TimeSinceCleaning", "3", "9"),
                   alpha = alpha, lfcThreshold = 0, altHypothesis = "greaterAbs")
  res69 <- results(dds, contrast = c("TimeSinceCleaning", "6", "9"),
                   alpha = alpha, lfcThreshold = 0, altHypothesis = "greaterAbs")
  
  f <- function(res){
    mcols(res, use.names = TRUE)
    sigtab <- res[which(res$padj < alpha), ]
  }
  
  f2 <- function(sigtab){
    cbind(as(sigtab, "data.frame"), as(tax_table(ps)[rownames(sigtab), ], "matrix"))
  }
  
  sigtab03 <- f(res03)
  sigtab06 <- f(res06)
  sigtab09 <- f(res09)
  sigtab36 <- f(res36)
  sigtab39 <- f(res39)
  sigtab69 <- f(res69)
  
  if (nrow(sigtab03) > 0){
    sigtab03 <- f2(sigtab03)
    write.csv(sigtab03, file.path(path.out, "03.csv"))
  }
  if (nrow(sigtab06) > 0){
    sigtab06 <- f2(sigtab06) 
    write.csv(sigtab06, file.path(path.out, "06.csv"))
  }
  if (nrow(sigtab09) > 0){
    sigtab09 <- f2(sigtab09) 
    write.csv(sigtab09, file.path(path.out, "09.csv"))
  }
  if (nrow(sigtab36) > 0){
   sigtab36 <- f2(sigtab36) 
   write.csv(sigtab36, file.path(path.out, "36.csv"))
  }
  if (nrow(sigtab39) > 0){
    sigtab39 <- f2(sigtab39) 
    write.csv(sigtab39, file.path(path.out, "39.csv"))
  }
  if (nrow(sigtab69) > 0){
    sigtab69 <- f2(sigtab69) 
    write.csv(sigtab69, file.path(path.out, "69.csv"))
  }
  
}

```


## Define function for plotting 
```{r}
# Heat plot based on CheeseOutcome. This function will plot only the top 30 taxa
ChouHeat <- function(ps, samType, w, h, path.out) {
  my_palette <- colorRampPalette(c("red", "yellow", "green"))(n = 299)
  # defines the color breaks manually for a "skewed" color transition
  col_breaks = c(seq(-1,0,length=100),      # for red
                 seq(0.01,0.8,length=100),  # for yellow
                 seq(0.81,1,length=100))    # for green

  #ps <- ps %>% transform_sample_counts(function(x) x/sum(x) )  #get abundance in %
  ps <- ps %>% tax_glom(taxrank = "Genus")
  taxa.df <- psmelt(ps)
  taxa.agg <- aggregate(Abundance ~ Genus + CheeseOutcome,
                        data = taxa.df,
                        mean)
  taxa.cast <- dcast(taxa.agg, Genus ~ CheeseOutcome, mean, value.var = "Abundance")
  # need to change results from factor to numeric because of R
  row.names(taxa.cast) <- taxa.cast$Genus
  taxa.cast <- taxa.cast[, -1]
  indx <- sapply(taxa.cast, is.factor)
  taxa.cast[indx] <- lapply(taxa.cast[indx], function(x) as.numeric(as.character(x))) 
  taxa.cast30 <- cbind(taxa.cast, total = rowSums(taxa.cast)) #  need numeric values
  taxa.cast30$taxa <- rownames(taxa.cast30)
  taxa.cast30 <- head(arrange(taxa.cast30,desc(total)), n = 30)
  row.names(taxa.cast30) <- taxa.cast30$taxa
  taxa.cast30 <- taxa.cast30[, -c(3,4)]
  
  pdf(path.out, width = w, height = h)
  superheat(taxa.cast30,
            left.label.size = 0.4, 
            bottom.label.size = 0.1,
            order.rows = rev(order(rownames(taxa.cast30))),
            #order.cols = c("1",2,3), # follow raw --> feed --> htst milk
            grid.hline = FALSE,
            title = samType,
            title.size = 6,
            title.alignment = "center")
  dev.off()
}


# Heat plot based on TimeSinceCleaning. This function will plot only the top 30 taxa
TiSiHeat <- function(ps, samType, w, h, path.out) {
  my_palette <- colorRampPalette(c("red", "yellow", "green"))(n = 299)
  # defines the color breaks manually for a "skewed" color transition
  col_breaks = c(seq(-1,0,length=100),      # for red
                 seq(0.01,0.8,length=100),  # for yellow
                 seq(0.81,1,length=100))    # for green

  #ps <- ps %>% transform_sample_counts(function(x) x/sum(x) )  #get abundance in %
  ps <- ps %>% tax_glom(taxrank = "Genus")
  taxa.df <- psmelt(ps)
  taxa.agg <- aggregate(Abundance ~ Genus + TimeSinceCleaning,
                        data = taxa.df,
                        mean)
  taxa.cast <- dcast(taxa.agg, Genus ~ TimeSinceCleaning, mean, value.var = "Abundance")
  # need to change results from factor to numeric because of R
  row.names(taxa.cast) <- taxa.cast$Genus
  taxa.cast <- taxa.cast[, -1]
  indx <- sapply(taxa.cast, is.factor)
  taxa.cast[indx] <- lapply(taxa.cast[indx], function(x) as.numeric(as.character(x))) 
  taxa.cast30 <- cbind(taxa.cast, total = rowSums(taxa.cast)) #  need numeric values
  taxa.cast30$taxa <- rownames(taxa.cast30)
  taxa.cast30 <- head(arrange(taxa.cast30,desc(total)), n = 30)
  row.names(taxa.cast30) <- taxa.cast30$taxa
  taxa.cast30 <- taxa.cast30[, -c(5,6)]
  
  pdf(path.out, width = w, height = h)
  superheat(taxa.cast30,
            left.label.size = 0.4, 
            bottom.label.size = 0.1,
            order.rows = rev(order(rownames(taxa.cast30))),
            #order.cols = c("1",2,3), # follow raw --> feed --> htst milk
            grid.hline = FALSE,
            title = samType,
            title.size = 6,
            title.alignment = "center")
  dev.off()
}


# Stacked bar plot 
ChouBar <- function(ps, taxa, samType, w, h, path.out){
  ps <- ps %>% tax_glom(taxrank = "Genus") #%>%transform_sample_counts(function(x) x/sum(x) ) 
  TaxTab <- tax_table(ps) %>% as.data.frame()
  taxa_names(ps) <- TaxTab$Genus
  allTaxa <- taxa_names(ps)
  ps.notaxa <- prune_taxa(allTaxa[!(allTaxa %in% taxa)], ps)
  taxa.filt <- ps.notaxa@tax_table[,6] %>% as.character() #6 for genus level 
  
  # change the taxa that are not within the 13 common taxa to "Other" 
  TaxTab2 <- psmelt(ps)
  # merge/average per SampleType and CheeseOutcome
  TaxTab2_agg <- aggregate(Abundance ~ SampleType + CheeseOutcome + Genus,
                           data = TaxTab2,
                           mean)
  TaxTab2_agg$Genus <- as.character(TaxTab2_agg$Genus)
  TaxTab2_agg[TaxTab2_agg$Genus %in% taxa.filt,]$Genus <- "Other"

  #Set colors for plotting
  mycol = c("#A6CEE3", "#438EC0", "#63A8A0", "#98D277", "#3BA432", "#F16667", 
            "#E62F27", "#F9A963", "#FE982C", "#ED8F47", "#C3AAD2", "#7D54A5", 
            "#B9A499", "#EAD27A", "#B15928")
  TaxTab2_agg$Genus = factor(TaxTab2_agg$Genus, levels = taxa)
  
  #pdf(path.out, width = w, height = h)
  ggplot(TaxTab2_agg, aes(x = CheeseOutcome, y = Abundance, fill = Genus)) +
    facet_wrap(~SampleType, scales = "fixed") +
    geom_bar(stat = "identity") + 
    scale_fill_manual(values = mycol)+
    # Remove x axis title
    theme(axis.title.x = element_blank()) + 
    guides(fill = guide_legend(reverse = FALSE, keywidth = 1, keyheight = 1)) +
    ylab("Relative Abundance \n") 
  #dev.off()
}


# Stacked bar plot for time since cleaning
TiSinBar <- function(ps, taxa, samType, w, h, path.out){
  ps <- ps %>% tax_glom(taxrank = "Genus") #%>%transform_sample_counts(function(x) x/sum(x) ) 
  TaxTab <- tax_table(ps) %>% as.data.frame()
  taxa_names(ps) <- TaxTab$Genus
  allTaxa <- taxa_names(ps)
  ps.notaxa <- prune_taxa(allTaxa[!(allTaxa %in% taxa)], ps)
  taxa.filt <- ps.notaxa@tax_table[,6] %>% as.character() #6 for genus level 
  
  # change the taxa that are not within the 13 common taxa to "Other" 
  TaxTab2 <- psmelt(ps)
  # merge/average per SampleType and TimeSinceCleaning
  TaxTab2_agg <- aggregate(Abundance ~ SampleType + TimeSinceCleaning + Genus,
                           data = TaxTab2,
                           mean)
  TaxTab2_agg$Genus <- as.character(TaxTab2_agg$Genus)
  TaxTab2_agg[TaxTab2_agg$Genus %in% taxa.filt,]$Genus <- "Other"

  #Set colors for plotting
  mycol = c("#A6CEE3", "#438EC0", "#63A8A0", "#98D277", "#3BA432", "#F16667", 
            "#E62F27", "#F9A963", "#FE982C", "#ED8F47", "#C3AAD2", "#7D54A5", 
            "#B9A499", "#EAD27A", "#B15928")
  TaxTab2_agg$Genus = factor(TaxTab2_agg$Genus, levels = taxa)
  
  #pdf(path.out, width = w, height = h)
  ggplot(TaxTab2_agg, aes(x = TimeSinceCleaning, y = Abundance, fill = Genus)) +
    facet_wrap(~SampleType, scales = "fixed") +
    geom_bar(stat = "identity") + 
    scale_fill_manual(values = mycol)+
    # Remove x axis title
    theme(axis.title.x = element_blank()) + 
    guides(fill = guide_legend(reverse = FALSE, keywidth = 1, keyheight = 1)) +
    ylab("Relative Abundance \n") 
  #dev.off()
}
```


## Prepare phyloseq objects 
```{r}
# remake a unrarefied ps object with low coverage samples filtered
## The starting ps is removed of ASV with undetermined phylum and ASV < 0.00005
ps05.Phy05 <- readRDS(file.path(path, "decontam/phyloseq/ps05.Phy05.rds"))
ps05.Phy05
# add tree
tree <- read_tree(file.path(path, "decontam/phyloseq/tree.nwk"))
ps05.Phy05.tre <- merge_phyloseq(ps05.Phy05, phy_tree(tree))
saveRDS(ps05.Phy05.tre, file.path(path, "decontam/phyloseq/ps05.Phy05.tre.rds"))
ps05.Phy05.tre

# remove sample < 100 reads
ps100 <- prune_samples(sample_sums(ps05.Phy05.tre) >= 100, ps05.Phy05.tre)
saveRDS(ps100, file.path(path, "decontam/phyloseq/ps100.rds"))
ps100

# remove all NC and seq control samples 
ps100.sam <- subset_samples(ps100, SampleOrCtrl == "Sample" )
#ps100.sam <- ps100.sam %>% tax_glom(taxrank = "Genus")
saveRDS(ps100.sam, file.path(path, "decontam/phyloseq/ps100.sam.rds"))
ps100.sam

# Feed and HTST milk with information on TimeSinceCleaning
ps.milk <- subset_samples(ps100.sam, SampleType %in% c("HTST_milk","HTST_feed"))
ps.milk
## Read in new sample information dataframe 
sam.milk <- read.csv(file.path(path, "mapping/sam_milk_HTST0369.csv"))
rownames(sam.milk) <- sam.milk$SampleID
## Merge as a new phyloseq object
tree <- phy_tree(ps.milk)
tax <- tax_table(ps.milk)
otu <- otu_table(ps.milk)
ps.milk <- phyloseq(otu_table(otu, taxa_are_rows = FALSE), tax_table(tax),sample_data(sam.milk), phy_tree(tree))
saveRDS(ps.milk, file.path(path, "milk/ps.milk.rds"))
ps.milk

# Feed and HTST milk that are paired with cheddar (based on the 4hr timeline)
samdf.4hr <- read.csv(file.path(path, "mapping/Run1-5_samdf-mc2_4hrTL.csv"))
rownames(samdf.4hr) <- samdf.4hr$SampleID
ps.milk4 <- phyloseq(otu_table(otu, taxa_are_rows = FALSE), tax_table(tax),sample_data(samdf.4hr), phy_tree(tree))
saveRDS(ps.milk4, file.path(path, "milk/ps.milk4.rds"))
ps.milk4
```


## Taxonomic DA of Cheddar 
This is PMA treated samples
```{r}
# 1 # All Cheddar samples 
ps.ched <- subset_samples(ps100.sam, CheeseType %in% "Cheddar")
ps.ched <- subset_samples(ps.ched, 
                          SampleType %in% c("Cheese_0D","Cheese_30D","Cheese_90D","Cheese_120D"))
saveRDS(ps.ched, file.path(path, "cheddar/ps.ched.rds"))
ps.ched
# PMA 
ps.chedY <- subset_samples(ps.ched, PMA %in% "Y")
saveRDS(ps.chedY, file.path(path, "cheddar/ps.chedY.rds"))
ps.chedY
CheeseOutcomeDA(ps.chedY, path.out = file.path(path, "cheddar/deseq_table/sigtab.chedY.csv"))

# 2 # All Cheddar samples without Lactococcus
ps.chedY.noL <- subset_taxa(ps.chedY, !Genus %in% "Lactococcus")
saveRDS(ps.chedY.noL, file.path(path, "cheddar/ps.chedY.noL.rds"))
ps.chedY.noL
CheeseOutcomeDA(ps.chedY.noL, 
                path.out = file.path(path, "cheddar/deseq_table/sigtab.chedY.noL.csv"))

# 3 # 0-day Cheddar samples without Lactococcus
ps.chedY.noL0 <- subset_samples(ps.chedY.noL, SampleType %in% "Cheese_0D")
saveRDS(ps.chedY.noL0, file.path(path, "cheddar/ps.chedY.noL0.rds"))
ps.chedY.noL0
CheeseOutcomeDA(ps.chedY.noL0, 
                path.out = file.path(path, "cheddar/deseq_table/sigtab.chedY.noL0.csv"))

# 4 # 30-day Cheddar samples without Lactococcus
ps.chedY.noL30 <- subset_samples(ps.chedY.noL, SampleType %in% "Cheese_30D")
saveRDS(ps.chedY.noL30, file.path(path, "cheddar/ps.chedY.noL30.rds"))
ps.chedY.noL30
CheeseOutcomeDA(ps.chedY.noL30, 
                path.out = file.path(path, "cheddar/deseq_table/sigtab.chedY.noL30.csv"))

# 5 # 90-day Cheddar samples without Lactococcus
ps.chedY.noL90 <- subset_samples(ps.chedY.noL, SampleType %in% "Cheese_90D")
saveRDS(ps.chedY.noL90, file.path(path, "cheddar/ps.chedY.noL90.rds"))
ps.chedY.noL90
CheeseOutcomeDA(ps.chedY.noL90, 
                path.out = file.path(path, "cheddar/deseq_table/sigtab.chedY.noL90.csv"))

# 6 # 120-day Cheddar samples without Lactococcus
ps.chedY.noL120 <- subset_samples(ps.chedY.noL, SampleType %in% "Cheese_120D")
saveRDS(ps.chedY.noL120, file.path(path, "cheddar/ps.chedY.noL120.rds"))
ps.chedY.noL120
CheeseOutcomeDA(ps.chedY.noL120, 
                path.out = file.path(path, "cheddar/deseq_table/sigtab.chedY.noL120.csv"))
```


## Taxonomic DA of HTST milk that has not been adjusted for cell count using qPCR results
This is PMA treated samples
```{r}
# 7 # HTST milk 
ps.htst <- subset_samples(ps.milk, SampleType %in% "HTST_milk")
saveRDS(ps.htst, file.path(path, "milk/ps.htst.rds"))
ps.htst
ps.htstY <- subset_samples(ps.htst, PMA %in% "Y")
saveRDS(ps.htstY, file.path(path, "milk/ps.htstY.rds"))
ps.htstY
CheeseOutcomeDA(ps.htstY, 
                path.out = file.path(path, "milk/deseq_table/sigtab.htstY.csv"))
ChouHeat(ps.htstY, samType = "PMA treated HTST milk", w = 8.5, h = 11,
         path.out = file.path(path, "milk/fig/htstY.pdf"))
TiSinCleaningDA(ps.htstY, 
                path.out = file.path(path, "milk/deseq_table_TiSinCleaning/htstY"))
TiSiHeat(ps.htstY, samType = "PMA treated HTST milk", w = 8.5, h = 11,
         path.out = file.path(path, "milk/fig_TiSinCleaning/htstY.pdf"))


# 8 # HTST milk that are paired with cheese (4hr timeline)
ps.htst4 <- subset_samples(ps.milk4, SampleType %in% "HTST_milk")
saveRDS(ps.htst4, file.path(path, "milk/ps.htst4.rds"))
ps.htst4
ps.htst4Y <- subset_samples(ps.htst4, PMA %in% "Y")
saveRDS(ps.htstY, file.path(path, "milk/ps.htstY.rds"))
ps.htst4Y
CheeseOutcomeDA(ps.htst4Y, 
                path.out = file.path(path, "milk/deseq_table/sigtab.htst4Y.csv"))
ChouHeat(ps.htst4Y, samType = "PMA treated paired HTST milk", w = 8.5, h = 11,
         path.out = file.path(path, "milk/fig/htst4Y.pdf"))


```


## Taxonomic DA of HTST milk that ARE adjusted for cell count using qPCR results
DESeq2 keeps giving error because after qpcr adjustment, it is not raw counts anymore.
```{r eval=FALSE, include=FALSE}
qPCR <- read.csv(file.path(path, "mapping/Milk_sample_qPCR.csv"))
# transform to percentage
ps.milkq <- ps.milk %>% 
  tax_glom(taxrank = "Genus") %>% 
  transform_sample_counts(function(x) x/sum(x) )  
otuq <- otu_table(ps.milkq) %>% data.frame()
otuq$Sampleid <- row.names(otuq)
colnames(otuq) <- gsub("X","",colnames(otuq))
otu.q <- merge(x = qPCR, y = otuq, by.x = "SampleID", by.y = "Sampleid")
coln <- colnames(otu.q[,c(20:91)]) #  character vector with all feature names
otu.q2 <- otu.q %>% mutate_at(.vars =  coln, .funs = funs(.*LogCellsML))
row.names(otu.q2) <- otu.q2[,1] #sample id as row names for formatting with phyloseq
otu.q3 <- otu.q2[,-c(1:19)] #remove all the meta information so the resulting otu table can be transformed into a numeric matrix
otu.q3 <- as.matrix(otu.q3)

# Merge as a new phyloseq object
ps.milkq <- phyloseq(otu_table(otu.q3, taxa_are_rows = FALSE), 
                     tax_table(ps.milkq),
                     sample_data(ps.milkq), 
                     phy_tree(ps.milkq))
# remove features that sum up to 0
#ps.milkq <- filter_taxa(ps.milkq, function(x) sum(x) > 0, TRUE)
#ps.milkq <- prune_samples(sample_sums(ps.milkq)>=0, ps.milkq)
saveRDS(ps.milkq, file.path(path, "milk/ps.milkq.rds"))
ps.milkq



# 9 # HTST milk that has been adjusted with qPCR values
ps.htstq <- subset_samples(ps.milkq, SampleType %in% "HTST_milk")
saveRDS(ps.htstq, file.path(path, "milk/ps.htstq.rds"))
ps.htstq
ps.htstqY <- subset_samples(ps.htstq, PMA %in% "Y")
saveRDS(ps.htstqY, file.path(path, "milk/ps.htstqY.rds"))
ps.htstqY
CheeseOutcomeDA(ps.htstqY, 
                path.out = file.path(path, "milk/deseq_table/sigtab.htstqY.csv"))
ChouBar(ps.htstqY, 
        taxa = comm14,
        samType = "PMA treated HTST milk", w = 6, h = 5,
        path.out = file.path(path, "milk/fig/htstqY2.pdf"))
TiSinCleaningDA(ps.htstqY, 
                path.out = file.path(path, "milk/deseq_table_TiSinCleaning/sigtab.htstqY.csv"))
TiSinBar(ps.htstqY, 
        taxa = comm14,
        samType = "PMA treated HTST milk", w = 6, h = 5,
        path.out = file.path(path, "milk/fig_TiSinCleaning/htstqY2.pdf"))


# 10 # HTST milk around the 4hr timeline that has been adjusted with qPCR values
CheeseOutcomeDA(ps.htstqY, 
                path.out = file.path(path, "milk/deseq_table/sigtab.htstqY.csv"))
ChouBar(ps.htstqY, 
        taxa = comm14,
        samType = "PMA treated HTST milk", w = 6, h = 5,
        path.out = file.path(path, "milk/fig/htstqY2.pdf"))

```


## Plotting and saving as pdf
```{r}
ps.milkY <- subset_samples(ps.milk, PMA %in% "Y")
ChouBar(ps.milkY, 
        taxa = comm14,
        samType = "PMA treated HTST milk", w = 6, h = 5,
        path.out = file.path(path, "milk/fig/milkY.pdf"))


ps.milkqY <- subset_samples(ps.milkq, PMA %in% "Y")
ChouBar(ps.milkqY, 
        taxa = comm14,
        samType = "PMA treated HTST milk", w = 6, h = 5,
        path.out = file.path(path, "milk/fig/milkqY.pdf"))
TiSinBar(ps.milkqY, 
        taxa = comm14,
        samType = "PMA treated HTST milk", w = 6, h = 5,
        path.out = file.path(path, "milk/fig_TiSinCleaning/htstqY2.pdf"))



ps.milkq4 <- phyloseq(otu_table(ps.milkq), tax_table(ps.milkq),sample_data(samdf.4hr), phy_tree(ps.milkq))
ps.milkq4Y <- subset_samples(ps.milkq4, PMA %in% "Y")
ChouBar(ps.milkq4Y, 
        taxa = comm14,
        samType = "PMA treated HTST milk", w = 6, h = 5,
        path.out = file.path(path, "milk/fig/milkq4Y.pdf"))

```

