---
title: "Cheese_paper_study1_portion"
author: "Zeya Xue"
date: "3/11/2019"
output: rmarkdown::github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.path = "figs/")
```

## Setting up packages and working directory 
```{r}
library(phyloseq);packageVersion("phyloseq")
library(tidyr)
library(ggplot2)
library(reshape2)
library(plyr)
library(dplyr)
library(PMCMR)
library(vegan)
library(RColorBrewer)
library(superheat)
library(DESeq2);packageVersion("DESeq2")

path <- "~/Google Drive File Stream/My Drive/UC_Davis/Marco_lab/milk_microbiota/Hilmar_weekly_samples/QIIME2/"

# read in ps object after initial organizatio 
ps <- readRDS(file.path(path, "decontam/phyloseq/ps05.rds"))
ps # 1646 taxa and 614 samples
tree <- read_tree(file.path(path,"decontam/phyloseq/tree.nwk"))
ps <- merge_phyloseq(ps, phy_tree(tree))
ps # 1646 taxa and 614 samples
# Remove species level 
ps <- phyloseq(otu_table(ps), sample_data(ps), phy_tree(ps),
               tax_table(as.matrix(tax_table(ps))[,c(-7,-8)]))
# remove sample < 500 reads
ps <- prune_samples(sample_sums(ps) >= 500, ps)
ps # 1646 taxa and 587 samples 
# Remove Archaea Chloroplast and mitochondria
ps <- subset_taxa(ps, !Family %in% "mitochondria") %>% 
  subset_taxa(!Kingdom %in% "Archaea") %>% 
  subset_taxa(!Class %in% "Chloroplast")
ps # 1625 taxa and 587 samples

comm17 <- c("Acinetobacter", "Bacillus", "Bacillaceae", "Clostridium", 
            "Clostridiaceae", "Clostridiales", "Enterobacteriaceae", "Lactococcus", 
            "Lactobacillus", "Lactobacillaceae","Leuconostoc", "Pseudomonas",
            "Staphylococcus", "Streptococcus", "Thermus", "Turicibacter", "Other")
comm16 <- c("Acinetobacter", "Bacillus", "Bacillaceae", "Clostridium", 
            "Clostridiaceae", "Clostridiales", "Enterobacteriaceae", 
            "Lactobacillus", "Lactobacillaceae","Leuconostoc", "Pseudomonas",
            "Staphylococcus", "Streptococcus", "Thermus", "Turicibacter", "Other")

col17 <- c("#A6CEE3", "#438EC0", "#1f78b4","#b3e2cd", 
           "#98D277", "#3BA432", "#B89B74", "#E62F27", 
           "#F16667", "#fbb4ae", "#F9A963", "#ff7f00", 
           "#C3AAD2", "#7D54A5", "#B9A499", "#EAD27A", "#B15928")
col16 <- c("#A6CEE3", "#438EC0", "#1f78b4","#b3e2cd", 
           "#98D277", "#3BA432", "#B89B74", 
           "#F16667", "#fbb4ae", "#F9A963", "#ff7f00", 
           "#C3AAD2", "#7D54A5", "#B9A499", "#EAD27A", "#B15928")
```

## Filter ASVs based on taxonomy and abundance 
```{r}
# Show available ranks in the dataset
rank_names(ps)
# Create table, number of features for each phylum
table(tax_table(ps)[, "Phylum"], exclude = NULL)
# remove asv that was not assigned at phylum level
ps.Phy <- subset_taxa(ps, !is.na(Phylum) & !Phylum %in% c(""))
ps.Phy # 1493 taxa and 587 samples
# Explore abundance to decide filter threshold
## Compute prevalence of each feature, store as data.frame
prevdf <- apply(X = otu_table(ps.Phy),
                MARGIN = ifelse(taxa_are_rows(ps.Phy), yes = 1, no = 2),
                FUN = function(x)(sum(x>0)))
# Add taxonomy and total read counts to this data.frame
prevdf <- data.frame(Prevalence = prevdf,
                     TotalAbundance = taxa_sums(ps.Phy),
                     tax_table(ps.Phy))
# plot the abudance per phylum 
ggplot(prevdf, aes(TotalAbundance, Prevalence / nsamples(ps.Phy),color=Phylum)) +
  # Include a guess for parameter
  geom_hline(yintercept = 0.01,  linetype = 2) + 
  geom_point(size = 2, alpha = 0.7) +
  scale_x_log10() +  
  xlab("Total Abundance") + 
  ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~Phylum) + theme(legend.position="none")
# Remove very low abudance ASVs independent of sample prevalence
x = taxa_sums(ps.Phy)
keepTaxa = which((x / sum(x)) > 0.00005)
ps.Phy05 <- prune_taxa(names(keepTaxa), ps.Phy)
ps.Phy05 # 256 taxa and 587 samples
```

## Divide the ps object for different dataset 
### For taxonomy analysis based on "ps.REC.glom"
I only need to redo the taxa plot because the Genus level "gloming"
is only in taxonomy analysis, not alpha nor beta.
```{r}
# Clean up the taxonomy level
#make the deepest taxonomic identification in TaxTab as the recommened taxonomy (REC)
RECps <- dget("~/src/Marco_Lab/FUN_RECps.R") # Read in RECps function
ps.Phy.REC <- RECps(ps.Phy05@tax_table, ps.Phy05)
ps.Phy.REC # 256 taxa and 587 samples
ps.REC.glom <- tax_glom(ps.Phy.REC, taxrank = "REC", NArm = FALSE)
ps.REC.glom #  94 taxa and 587 samples


############################# 1. All milk samples #############################
ps.milk <- subset_samples(ps.REC.glom, SampleType %in% c("HTST_milk","Raw_milk","HTST_feed"))
ps.milk # 94 taxa and 376 samples 
# PMA treated sample
ps.milkY <- subset_samples(ps.milk, PMA %in% "Y")
ps.milkY # 94 taxa and 188 samples
# PMA untreated samples
ps.milkN <- subset_samples(ps.milk, PMA %in% "N")
ps.milkN # 94 taxa and 188 samples

################# 2. HTST feed and HTST milk samples after qPCR ################
qPCR <- read.csv(file.path(path, "mapping/Milk_sample_qPCR.csv"))
# transform to percentage
otuq <- ps.milk %>% transform_sample_counts(function(x) x/sum(x) )  %>% 
  otu_table() %>% data.frame()
otuq$Sampleid <- row.names(otuq)
colnames(otuq) <- gsub("X","",colnames(otuq))
otu.q <- merge(x = qPCR, y = otuq, by.x = "SampleID", by.y = "Sampleid")
coln <- colnames(otu.q[,c(20:91)]) #  character vector with all feature names
otu.q2 <- otu.q %>% mutate_at(.vars =  coln, .funs = funs(.*LogCellsML))
row.names(otu.q2) <- otu.q2[,1] #sample id as row names for formatting with phyloseq
otu.q3 <- otu.q2[,-c(1:19)] #remove all the meta information so the resulting otu table can be transformed into a numeric matrix
otu.q3 <- as.matrix(otu.q3)

# new sample information dataframe 
sam.milk <- read.csv(file.path(path, "mapping/sam_milk_HTST0369.csv"))
rownames(sam.milk) <- sam.milk$SampleID

# Merge as a new phyloseq object
ps.milkq <- phyloseq(otu_table(otu.q3, taxa_are_rows = FALSE), 
                     tax_table(ps.milk),
                     sample_data(sam.milk), 
                     phy_tree(ps.milk))
ps.milkq # 94 taxa and 262 samples 
ps.milkqY <- subset_samples(ps.milkq, PMA %in% "Y")
ps.milkqY # 94 taxa and 129 samples 
ps.milkqN <- subset_samples(ps.milkq, PMA %in% "N")
ps.milkqN # 94 taxa and 133 samples

################################ 3. Cheddar cheese ############################
ps.cheese <- subset_samples(ps.REC.glom, SampleType %in% c("Cheese_0D","Cheese_120D","Cheese_30D","Cheese_90D"))
ps.cheese # 94 taxa and 144 samples
ps.Ched <- subset_samples(ps.cheese, CheeseType %in% "Cheddar")
ps.Ched # 94 taxa and 110 samples 
ps.ChedY <- subset_samples(ps.Ched, PMA %in% "Y")
ps.ChedY # 94 taxa and 55 samples 
ps.ChedY.noL <- subset_taxa(ps.ChedY, !Genus %in% "Lactococcus")
ps.ChedY.noL # 93 taxa and 55 samples 
ps.ChedN <- subset_samples(ps.Ched, PMA %in% "N")
ps.ChedN # 94 taxa and 110 samples
ps.ChedN.noL <- subset_taxa(ps.ChedN, !Genus %in% "Lactococcus")
ps.ChedN.noL # 93 taxa and 55 samples 

######## 4. 4hr time line HTST feed and HTST milk samples after qPCR ###########
samdf.4hr <- read.csv(file.path(path, "mapping/Run1-5_samdf-mc2_4hrTL-NoExtremeMilk.csv"))
rownames(samdf.4hr) <- samdf.4hr$SampleID
ps.milkq4 <- phyloseq(otu_table(ps.milkq), tax_table(ps.milkq),
                      sample_data(samdf.4hr), phy_tree(ps.milkq))
ps.milkq4 # 94 taxa and 42 samples
ps.milkq4Y <- subset_samples(ps.milkq4, PMA %in% "Y")
ps.milkq4Y # 94  taxa and 21 samples
ps.milkq4N <- subset_samples(ps.milkq4, PMA %in% "N")
ps.milkq4N #  94 taxa and 21 samples

############# 4. 4hr time line HTST feed and HTST milk samples ################
ps.milk4 <- phyloseq(otu_table(ps.milk), tax_table(ps.milk),
                     sample_data(samdf.4hr), phy_tree(ps.milk))
ps.milk4 # 94 taxa and 48 samples
ps.milk4Y <- subset_samples(ps.milk4, PMA %in% "Y")
ps.milk4Y # 94  taxa and 24 samples


```


## Heat map figures
```{r}
# Heat plot based on MilkType
MilkHeat <- function(ps, taxa, samType, w, h, path.out) {
  my_palette <- colorRampPalette(c("red", "yellow", "green"))(n = 299)
  # defines the color breaks manually for a "skewed" color transition
  col_breaks = c(seq(-1,0,length=100),      # for red
                 seq(0.01,0.8,length=100),  # for yellow
                 seq(0.81,1,length=100))    # for green
  
  ps <- ps %>% transform_sample_counts(function(x) x/sum(x) )  
  taxa.df <- psmelt(ps)
  taxa.agg <- aggregate(Abundance ~ REC + SampleType,
                        data = taxa.df,
                        mean)
  taxa.cast <- dcast(taxa.agg, REC ~ SampleType, mean, value.var = "Abundance")
  # Make filter vector for the 14 selected cheddar taxa 
  TaxTab <- tax_table(ps) %>% as.data.frame()
  taxa_names(ps) <- TaxTab$REC
  allTaxa <- taxa_names(ps)
  ps.notaxa <- prune_taxa(allTaxa[!(allTaxa %in% taxa)], ps)
  taxa.filt <- ps.notaxa@tax_table[,7] %>% as.character() # 7 for REC level 
  # Filter
  taxa.cast[taxa.cast$REC %in% taxa.filt,]$REC <- "Other"
  taxa.cast.f <- taxa.cast[!(is.na(taxa.cast$REC) | taxa.cast$REC==""), ]
  
  # make as numeric matrix 
  row.names(taxa.cast.f) <- taxa.cast.f$REC
  taxa.cast.f <- taxa.cast.f[,-1]
  
  png(path.out, width = w, height = h)
  (superheat(taxa.cast.f,
            left.label.size = 0.7, 
            bottom.label.size = 0.1,
            order.rows = rev(order(rownames(taxa.cast.f))),
            order.cols = c(3,1,2), # follow raw --> feed --> htst milk
            #col.dendrogram = TRUE,
            grid.hline = FALSE,
            title = samType,
            title.size = 6,
            title.alignment = "center")) %>% print()
  dev.off()
}


MilkHeat(ps.milkY, taxa = comm17, 
         samType = "PMA treated milk", w = 450, h = 600,
         path.out = file.path(path, "Mar_2019/Heat_sampletype_milkY.png"))
MilkHeat(ps.milkN, comm17, 
         samType = "PMA untreated milk", w = 450, h = 600,
         path.out = file.path(path, "Mar_2019/Heat_sampletype_milkN.png"))

```


## Line plot 
```{r}
# Define function for line plot
LinePlot <- function(ps, x, w, h, path.out){
  melt <- psmelt(ps) 
  agg <- ddply(melt, c("TimeSinceCleaning","SampleType","PMA", "REC"),
               summarise,
               N = length(Abundance),
               mean = mean(Abundance),
               sd = sd(Abundance),
               se = sd / sqrt(N))
  
  pdf(path.out, width = w, height = h)
  print(ggplot(subset(agg, REC %in% x), aes(x = TimeSinceCleaning, y = mean, color = SampleType)) +
              geom_line(size = 0.9)+
              geom_point(size = 1)+
              geom_errorbar(aes(ymin = mean-se, ymax = mean+se),width = 0.1, position=position_dodge(0.05))+
              ylab(x) +
              theme_bw() +
              theme(text = element_text(size = 15)))
  dev.off()
}


# loop for all specific taxa at REC level for PMA treated 
for(i in 1:length(tax_table(ps.milkq)[,7])){
  LinePlot(ps.milkqY, x = tax_table(ps.milkq)[,7][i], w = 4.5, h = 3.5, 
           path.out = file.path(path, "Mar_2019/line_tisinclean", 
                                paste(tax_table(ps.milkq)[,7][i],".pdf", sep ='')))
}
```


## Bar plot 
```{r}
Bar <- function(ps, var1, taxa, lvl1, mycol, path.out, w, h){  
  ps <- ps %>% transform_sample_counts(function(x) x/sum(x) ) 
  
  TaxTab <- tax_table(ps) %>% as.data.frame()
  taxa_names(ps) <- TaxTab$REC
  allTaxa <- taxa_names(ps)
  ps.other <- prune_taxa(allTaxa[!(allTaxa %in% taxa)], ps)
  taxa.other <- ps.other@tax_table[,7] %>% as.character() # 7 for REC level  
  
  # change the taxa that are not the expected taxa to "Other" 
  TaxTab2 <- psmelt(ps)
  # merge/average per Sample and CheeseOutcome
  TaxTab2_agg <- ddply(TaxTab2, c(var1, "REC"), 
                       summarise,
                       N = length(Abundance),
                       mean = mean(Abundance),
                       sd = sd(Abundance),
                       se = sd / sqrt(N))
  
  colnames(TaxTab2_agg)[1] <- var1
  TaxTab2_agg$REC <- as.character(TaxTab2_agg$REC)
  TaxTab2_agg[TaxTab2_agg$REC %in% taxa.other,]$REC <- "Other"
  
  # Set orders for taxonomy and sample
  TaxTab2_agg$REC <- factor(TaxTab2_agg$REC, levels = taxa)
  TaxTab2_agg[[var1]] <- factor(TaxTab2_agg[[var1]], levels = lvl1)
  
  pdf(path.out, w, h)
  print(ggplot(TaxTab2_agg, aes(x = get(var1), y = mean, fill = REC)) +
    geom_bar(stat = "identity",position = "dodge" ) + 
    geom_errorbar(aes(ymin = mean-se, ymax = mean+se), width = 0.8, position = position_dodge(.9)) +
    scale_fill_manual(values = mycol)+
    theme_bw()+
    theme(axis.text.x = element_text(angle = 45, hjust = 1))+
    theme(text = element_text(size = 15))+
    guides(fill = guide_legend(reverse = FALSE, keywidth = 1, keyheight = 1)) +
    ylab("Relative Abundance \n"))
  dev.off()
}

Bar(ps.ChedY.noL, var1 = "SampleType", 
    lvl1 = c("Cheese_0D","Cheese_30D","Cheese_90D","Cheese_120D"),
    taxa = comm16, mycol = col16,
    path.out = file.path(path, "Mar_2019/ChedY_noL_sampletype.pdf"),6.5,5.5)
Bar(ps.ChedN.noL, var1 = "SampleType", 
    lvl1 = c("Cheese_0D","Cheese_30D","Cheese_90D", "Cheese_120D"),
    taxa = comm16, mycol = col16,
    path.out = file.path(path, "Mar_2019/ChedN_noL_sampletype.pdf"),6.5, 5.5)
```



## Stacked bar plot
```{r}
ChouBar <- function(ps, taxa, mycol, samType,lvl, w, h, path.out){
  TaxTab <- tax_table(ps) %>% as.data.frame()
  taxa_names(ps) <- TaxTab$REC
  allTaxa <- taxa_names(ps)
  ps.notaxa <- prune_taxa(allTaxa[!(allTaxa %in% taxa)], ps)
  taxa.filt <- ps.notaxa@tax_table[,7] %>% as.character() # 7 for REC level 
  
  # change the taxa that are not within the 13 common taxa to "Other" 
  TaxTab2 <- psmelt(ps)
  # merge/average per SampleType and CheeseOutcome
  TaxTab2_agg <- ddply(TaxTab2, c("SampleType", "CheeseOutcome", "REC"),
                       summarise,
                       mean = mean(Abundance))
  TaxTab2_agg$REC <- as.character(TaxTab2_agg$REC)
  TaxTab2_agg[TaxTab2_agg$REC %in% taxa.filt,]$REC <- "Other"
  TaxTab2_agg$REC <- factor(TaxTab2_agg$REC, levels = taxa)
  TaxTab2_agg$SampleType <- factor(TaxTab2_agg$SampleType, levels = lvl)
  
  pdf(path.out, width = w, height = h)
  (ggplot(TaxTab2_agg, aes(x = CheeseOutcome, y = mean, fill = REC)) +
    facet_wrap(~ SampleType) +
    geom_bar(stat = "identity") + 
    scale_fill_manual(values = mycol)+
    theme_bw() +
    # Remove x axis title
    theme(axis.title.x = element_blank(), text = element_text(size = 16)) + 
    guides(fill = guide_legend(reverse = FALSE, keywidth = 1, keyheight = 1)) +
    ylab("Log cells/ml")) %>% print() 
  dev.off()
}


ChouBar(ps.milkq4Y, taxa = comm17, mycol = col17, lvl = c("HTST_feed", "HTST_milk"),
        samType = "PMA treated HTST milk", w = 6, h = 5,
        path.out = file.path(path, "Mar_2019/milkq4Y_cheeseOutcome.pdf"))
ChouBar(ps.milkq4N, taxa = comm17, mycol = col17, lvl = c("HTST_feed", "HTST_milk"),
        samType = "PMA untreated HTST milk", w = 6, h = 5,
        path.out = file.path(path, "Mar_2019/milkq4N_cheeseOutcome.pdf"))
ChouBar(ps.ChedY.noL %>% transform_sample_counts(function(x) x/sum(x) ) , 
        taxa = comm16, mycol = col16,
        lvl = c("Cheese_0D","Cheese_30D","Cheese_90D", "Cheese_120D"),
        samType = "PMA treated HTST milk", w = 6, h = 5,
        path.out = file.path(path, "Mar_2019/ChedY.noL_cheeseOutcome.pdf"))
# Present in actual relatvie propotion of non-starter bacteria
ChouBar(subset_taxa(ps.ChedY %>% transform_sample_counts(function(x) x/sum(x)) , !Genus %in% "Lactococcus"), 
        taxa = comm16, mycol = col16,
        lvl = c("Cheese_0D","Cheese_30D","Cheese_90D", "Cheese_120D"),
        samType = "PMA treated HTST milk", w = 6, h = 5,
        path.out = file.path(path, "Mar_2019/ChedY.noL_actualprop_cheeseOutcome.pdf"))
```


## DEseq 
```{r}
# calculate geometric means prior to estimate size factors
gmMean = function(x, na.rm=TRUE){
  exp(sum(log(x[x > 0]), na.rm=na.rm) / length(x))
}
# Cheese Outcome Differential abundance and generate diagnostic plots and write sig tables 
## fold change is slits vs no slits
## Uses default Benjamini-Hochberg pvalue adjust
CheeseOutcomeDA <- function(ps, path.out) {
  psdds <- phyloseq_to_deseq2(ps, ~ CheeseOutcome)
  
  geoMeans <- apply(counts(psdds), 1, gmMean)
  psdds <- estimateSizeFactors(psdds, geoMeans = geoMeans)
  
  dds <- DESeq(psdds, test = "Wald", fitType = "local") 
  plotDispEsts(dds, ylim = c(1e-6, 1e2)) 
  alpha <- 0.1 #  padj for indepenent filtering and expect FDR < alpha
  res <- results(dds, alpha = alpha, 
                 lfcThreshold = 0, 
                 altHypothesis = "greaterAbs")
  mcols(res, use.names=TRUE)  
  plotMA(res) # diagnostic description
  hist(res$pvalue, breaks=20, col="grey" ) # diagnostic plot
  sigtab = res[which(res$padj < alpha), ] 
  if (nrow(sigtab) == 0) {
    print("There are no significant features whose padj < 0.1 and |lfc| > 0")
  } else {
    sigtab <- cbind(as(sigtab, "data.frame"), as(tax_table(ps)[rownames(sigtab), ], "matrix"))
    write.csv(sigtab, file.path(path.out))
  }
}



CheeseOutcomeDA(subset_samples(ps.ChedY.noL, SampleType %in% "Cheese_0D"), 
                path.out = file.path(path, "Mar_2019/sigtab.chedY.csv"))
CheeseOutcomeDA(subset_samples(ps.ChedY.noL, SampleType %in% "Cheese_30D"), 
                path.out = file.path(path, "Mar_2019/sigtab.chedY30.csv"))
CheeseOutcomeDA(subset_samples(ps.ChedY.noL, SampleType %in% "Cheese_90D"), 
                path.out = file.path(path, "Mar_2019/sigtab.chedY90.csv"))
CheeseOutcomeDA(subset_samples(ps.ChedY.noL, SampleType %in% "Cheese_120D"), 
                path.out = file.path(path, "Mar_2019/sigtab.chedY120.csv"))

CheeseOutcomeDA(subset_samples(ps.milkY, SampleType %in% "HTST_feed"), 
                path.out = file.path(path, "Mar_2019/sigtab.HTST_feedY.csv"))
CheeseOutcomeDA(subset_samples(ps.milkY, SampleType %in% "HTST_milk"), 
                path.out = file.path(path, "Mar_2019/sigtab.HTST_milkY.csv"))

CheeseOutcomeDA(subset_samples(ps.milk4Y, SampleType %in% "HTST_feed"), 
                path.out = file.path(path, "Mar_2019/sigtab.HTST_feed4Y.csv"))
CheeseOutcomeDA(subset_samples(ps.milk4Y, SampleType %in% "HTST_milk"), 
                path.out = file.path(path, "Mar_2019/sigtab.HTST_milk4Y.csv"))


```


